{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "import copy\n",
    "import numpy as np\n",
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CAATC', 'AATCC', 'ATCCA', 'TCCAA', 'CCAAC']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Code Challenge: Solve the String Composition Problem.\n",
    "Input: An integer k and a string Text.\n",
    "Output: Compositionk(Text) (the k-mers can be provided in any order).\n",
    "'''\n",
    "\n",
    "def Composition(string, k):\n",
    "    composition = []\n",
    "    for start in range(len(string) - k +1):\n",
    "        end       = start + k\n",
    "        substring = string[start: end]\n",
    "        composition.append(substring)\n",
    "    return (composition)\n",
    "\n",
    "# Test\n",
    "k = 5\n",
    "string = 'CAATCCAAC'\n",
    "\n",
    "Composition(string, k)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ACCGAAGCT'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "String Spelled by a Genome Path Problem. Reconstruct a string from its genome path.\n",
    "Input: A sequence of k-mers Pattern1, … ,Patternn such that the last k - 1 symbols of Patterni are equal to the first k-1 symbolsof Patterni+1 for 1 ≤ i ≤ n-1.\n",
    "Output: A string Text of length k+n-1 such that the i-th k-mer in Text is equal to Patterni  (for 1 ≤ i ≤ n).\n",
    "'''\n",
    "\n",
    "def NaiveLink(patterns):\n",
    "    if type(patterns) == str:\n",
    "        patterns = patterns.split('\\n')\n",
    "    string = patterns[0]\n",
    "    for pattern in patterns[1:]:\n",
    "        string = string + pattern[-1]\n",
    "    return (string)\n",
    "\n",
    "# Test\n",
    "patterns = '''ACCGA\n",
    "CCGAA\n",
    "CGAAG\n",
    "GAAGC\n",
    "AAGCT'''\n",
    "\n",
    "NaiveLink(patterns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AGGCA': ['GGCAT', 'GGCAC'],\n",
       " 'CATGC': ['ATGCG'],\n",
       " 'GCATG': ['CATGC'],\n",
       " 'GGCAT': ['GCATG']}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Code Challenge: Solve the Overlap Graph Problem (restated below).\n",
    "Input: A collection Patterns of k-mers.\n",
    "Output: The overlap graph Overlap(Patterns), in the form of an adjacency list. \n",
    "'''\n",
    "\n",
    "def FindOverlap(patterns):\n",
    "    if type(patterns) == str:\n",
    "        patterns = patterns.split('\\n')\n",
    "    \n",
    "    adjacency_list = {}\n",
    "    prefixs = {}\n",
    "    suffixs = {}\n",
    "    for pattern in patterns:\n",
    "        suffixs[pattern] = pattern[1:]\n",
    "        if pattern[:-1] in prefixs:\n",
    "            prefixs[pattern[:-1]].append(pattern)\n",
    "        else:\n",
    "            prefixs[pattern[:-1]] = [pattern]\n",
    "    \n",
    "    for pattern,suffix in suffixs.items():\n",
    "        if suffix in prefixs:\n",
    "            adjacency_list[pattern] = prefixs[suffix]\n",
    "            \n",
    "    return(adjacency_list)\n",
    "\n",
    "# Test\n",
    "patterns = '''ATGCG\n",
    "GCATG\n",
    "CATGC\n",
    "AGGCA\n",
    "GGCAT\n",
    "GGCAC'''\n",
    "\n",
    "FindOverlap(patterns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AGG': ['GGG'],\n",
       " 'CAG': ['AGG', 'AGG'],\n",
       " 'GAG': ['AGG'],\n",
       " 'GGA': ['GAG'],\n",
       " 'GGG': ['GGG', 'GGA']}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "DeBruijn Graph from k-mers Problem: Construct the de Bruijn graph from a set of k-mers.\n",
    "Input: A collection of k-mers Patterns.\n",
    "Output: The adjacency list of the de Bruijn graph DeBruijn(Patterns).\n",
    "'''\n",
    "\n",
    "def PathGraph(patterns):\n",
    "    if type(patterns) == str:\n",
    "        patterns = patterns.split('\\n')\n",
    "    \n",
    "    graph = {}\n",
    "    for pattern in patterns:\n",
    "        if pattern[:-1] in graph:\n",
    "            graph[pattern[:-1]].append(pattern[1:])\n",
    "        else:\n",
    "            graph[pattern[:-1]] = [pattern[1:]]\n",
    "    return(graph)\n",
    "\n",
    "# Test\n",
    "patterns = '''GAGG\n",
    "CAGG\n",
    "GGGG\n",
    "GGGA\n",
    "CAGG\n",
    "AGGG\n",
    "GGAG'''\n",
    "\n",
    "PathGraph(patterns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AAG': ['AGA', 'AGA'],\n",
       " 'AGA': ['GAT'],\n",
       " 'ATT': ['TTC'],\n",
       " 'CTA': ['TAA'],\n",
       " 'CTC': ['TCT'],\n",
       " 'GAT': ['ATT'],\n",
       " 'TAA': ['AAG'],\n",
       " 'TCT': ['CTC', 'CTA'],\n",
       " 'TTC': ['TCT']}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Code Challenge: Solve the De Bruijn Graph from a String Problem.\n",
    "Input: An integer k and a string Text.\n",
    "Output: DeBruijnk(Text), in the form of an adjacency list.\n",
    "'''\n",
    "\n",
    "def DeBruijn(string, k):\n",
    "    composition  = Composition(string, k)\n",
    "    output_graph = PathGraph(composition)\n",
    "    \n",
    "    return (output_graph)\n",
    "\n",
    "# Test\n",
    "k = 4\n",
    "string = 'AAGATTCTCTAAGA'\n",
    "\n",
    "DeBruijn(string, k)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['6', '8', '7', '9', '6', '5', '4', '2', '1', '0', '3', '2', '6']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Code Challenge: Solve the Eulerian Cycle Problem.\n",
    "Input: The adjacency list of an Eulerian directed graph.\n",
    "Output: An Eulerian cycle in this graph.\n",
    "'''\n",
    "\n",
    "def GraphString2Dict(graph):\n",
    "    graph = graph.split('\\n')\n",
    "    graph_dict = {}\n",
    "    for path in graph:\n",
    "        path       = path.replace(' ','')\n",
    "        from_, to_ = path.split('->')\n",
    "        if ',' in to_:\n",
    "            to_              = to_.split(',')\n",
    "            graph_dict[from_] = to_\n",
    "        else:\n",
    "            graph_dict[from_] = [to_]\n",
    "    return(graph_dict)\n",
    "\n",
    "def EulerianCycle(graph):\n",
    "    if type(graph) == str:\n",
    "        graph = GraphString2Dict(graph)\n",
    "        \n",
    "    tmp_graph = copy.deepcopy(graph)\n",
    "    from_ = random.choice(list(tmp_graph.keys()))\n",
    "    cycle = [from_]\n",
    "    unexplored_nodes = {}\n",
    "    \n",
    "    while len(tmp_graph) != 0:\n",
    "        if from_ in tmp_graph:\n",
    "            to = random.choice(tmp_graph[from_])\n",
    "            cycle.append(to)\n",
    "            tmp_graph[from_].remove(to)\n",
    "            \n",
    "            if len(tmp_graph[from_]) == 0:\n",
    "                tmp_graph.pop(from_)\n",
    "                unexplored_nodes.pop(from_, None)\n",
    "            else:\n",
    "                unexplored_nodes[from_] = copy.deepcopy(tmp_graph[from_])\n",
    "            from_ = to\n",
    "        \n",
    "        else:\n",
    "            # find a random new start point from unexplored_nodes\n",
    "            unexplored_node_index = np.where(np.array(cycle) == random.choice(list(unexplored_nodes)))[0][-1]\n",
    "            unexplored_node       = cycle[unexplored_node_index]\n",
    "            next_node             = random.choice(unexplored_nodes[unexplored_node])\n",
    "            tmp_graph[unexplored_node].remove(next_node)\n",
    "            \n",
    "            # make the cycle go back to new start point\n",
    "            recover_cycle = cycle[unexplored_node_index:]\n",
    "            cycle         = cycle[:unexplored_node_index + 1]\n",
    "            from_         = next_node\n",
    "            cycle.append(from_)\n",
    "            \n",
    "            # reconstruct the tmp_graph and unexplored_nodes according to current cycle\n",
    "            unexplored_nodes = {}\n",
    "            tmp_graph        = copy.deepcopy(graph)\n",
    "            for i in range(len(cycle) - 1):\n",
    "                node    = cycle[i]\n",
    "                to_node = cycle[i + 1]\n",
    "                tmp_graph[node].remove(to_node)\n",
    "                if len(tmp_graph[node]) == 0:\n",
    "                    tmp_graph.pop(node)\n",
    "                    unexplored_nodes.pop(node, None)\n",
    "                else : \n",
    "                    unexplored_nodes[node] = copy.deepcopy(tmp_graph[node])\n",
    "    return(cycle)\n",
    "\n",
    "# Test\n",
    "graph = '''0 -> 3\n",
    "     1 -> 0\n",
    "     2 -> 1,6\n",
    "     3 -> 2\n",
    "     4 -> 2\n",
    "     5 -> 4\n",
    "     6 -> 5,8\n",
    "     7 -> 9\n",
    "     8 -> 7\n",
    "     9 -> 6'''\n",
    "\n",
    "EulerianCycle(graph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['6', '7', '8', '9', '6', '3', '0', '2', '1', '3', '4']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Code Challenge: Solve the Eulerian Path Problem.\n",
    "Input: The adjacency list of a directed graph that has an Eulerian path.\n",
    "Output: An Eulerian path in this graph.\n",
    "'''\n",
    "\n",
    "def CheckBalance(graph):\n",
    "    if type(graph) == str:\n",
    "        graph = GraphString2Dict(graph)\n",
    "    balance = {}\n",
    "    \n",
    "    for key, values in graph.items():\n",
    "        balance[key] = 0\n",
    "        for value in values:\n",
    "            balance[value] = 0\n",
    "    \n",
    "    for from_, tos in graph.items():\n",
    "        balance[from_] = balance[from_] + len(tos)\n",
    "        for to in tos:\n",
    "            balance[to] = balance[to] - 1\n",
    "    return (balance)\n",
    "\n",
    "def EulerianPath(graph):\n",
    "    if type(graph) == str:\n",
    "        graph = GraphString2Dict(graph)\n",
    "    \n",
    "    balance = CheckBalance(graph)\n",
    "    # find start point by balance > 0\n",
    "    for key, values in balance.items():\n",
    "        if values > 0 :\n",
    "            from_ = key\n",
    "            break\n",
    "            \n",
    "    tmp_graph = copy.deepcopy(graph)\n",
    "    cycle = [from_]\n",
    "    unexplored_nodes = {}\n",
    "    \n",
    "    while len(tmp_graph) != 0:\n",
    "        if from_ in tmp_graph:\n",
    "            to = random.choice(tmp_graph[from_])\n",
    "            cycle.append(to)\n",
    "            tmp_graph[from_].remove(to)\n",
    "            if len(tmp_graph[from_]) == 0:\n",
    "                tmp_graph.pop(from_)\n",
    "                unexplored_nodes.pop(from_, None)\n",
    "            else:\n",
    "                unexplored_nodes[from_] = copy.deepcopy(tmp_graph[from_])\n",
    "            from_ = to\n",
    "        \n",
    "        else:\n",
    "            unexplored_node_index = np.where(np.array(cycle) == random.choice(list(unexplored_nodes)))[0][-1]\n",
    "            unexplored_node       = cycle[unexplored_node_index]\n",
    "            next_node             = random.choice(unexplored_nodes[unexplored_node])\n",
    "            tmp_graph[unexplored_node].remove(next_node)\n",
    "\n",
    "            recover_cycle = cycle[unexplored_node_index:]\n",
    "            cycle         = cycle[:unexplored_node_index + 1]\n",
    "            from_         = next_node\n",
    "            cycle.append(from_)\n",
    "\n",
    "            unexplored_nodes = {}\n",
    "            tmp_graph        = copy.deepcopy(graph)\n",
    "            for i in range(len(cycle) - 1):\n",
    "                node    = cycle[i]\n",
    "                to_node = cycle[i + 1]\n",
    "                tmp_graph[node].remove(to_node)\n",
    "                if len(tmp_graph[node]) == 0:\n",
    "                    tmp_graph.pop(node)\n",
    "                    unexplored_nodes.pop(node, None)\n",
    "                else : \n",
    "                    unexplored_nodes[node] = copy.deepcopy(tmp_graph[node])\n",
    "    return(cycle)\n",
    "\n",
    "# Test\n",
    "graph='''0 -> 2\n",
    "     1 -> 3\n",
    "     2 -> 1\n",
    "     3 -> 0,4\n",
    "     6 -> 3,7\n",
    "     7 -> 8\n",
    "     8 -> 9\n",
    "     9 -> 6'''\n",
    "\n",
    "EulerianPath(graph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GGCTTACCA'"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Code Challenge: Solve the String Reconstruction Problem.\n",
    "Input: An integer k followed by a list of k-mers Patterns.\n",
    "Output: A string Text with k-mer composition equal to Patterns. (If multiple answers exist, you may return any one.)\n",
    "'''\n",
    "\n",
    "def Reconstruction(k, strings):\n",
    "    if type(strings) == str:\n",
    "        strings = strings.split('\\n')\n",
    "    graph          = PathGraph(strings)\n",
    "    cycle          = EulerianPath(graph)\n",
    "    reconstruction = NaiveLink(cycle)\n",
    "    return(reconstruction)\n",
    "\n",
    "# Test\n",
    "k = 4\n",
    "strings = '''CTTA\n",
    "ACCA\n",
    "TACC\n",
    "GGCT\n",
    "GCTT\n",
    "TTAC'''\n",
    "\n",
    "Reconstruction(k, strings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'1111011000010100'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Code Challenge: Solve the k-Universal Circular String Problem.\n",
    "Input: An integer k.\n",
    "Output: A k-universal circular string.\n",
    "'''\n",
    "\n",
    "def BinaryStrings(k):\n",
    "    strings = ['0','1']\n",
    "    for i in range(k - 1):\n",
    "        strings = strings + strings\n",
    "        half    = len(strings) / 2\n",
    "        for index in range(len(strings)):\n",
    "            if index < half:\n",
    "                strings[index] = strings[index] + '0'\n",
    "            else :\n",
    "                strings[index] = strings[index] + '1'\n",
    "    return(strings)\n",
    "\n",
    "def KUniversal(k):\n",
    "    strings        = BinaryStrings(k) \n",
    "    graph          = PathGraph(strings)\n",
    "    cycle          = EulerianCycle(graph)\n",
    "    reconstruction = NaiveLink(cycle)\n",
    "    # delete last k - 1 number, because it is a cycle \n",
    "    reconstruction = reconstruction[: -k + 1]\n",
    "    return(reconstruction)\n",
    "\n",
    "# Test\n",
    "k = 4\n",
    "\n",
    "KUniversal(k)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GTGGTCGTGAGATGTTGA'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Code Challenge: Solve the String Reconstruction from Read-Pairs Problem.\n",
    "Input: Integers k and d followed by a collection of paired k-mers PairedReads.\n",
    "Output: A string Text with (k, d)-mer composition equal to PairedReads.\n",
    "'''\n",
    "\n",
    "def PairedPathGraph(gapped_patterns, k):\n",
    "    if type(gapped_patterns) == str:\n",
    "        gapped_patterns = gapped_patterns.split('\\n')\n",
    "\n",
    "    graph = {}\n",
    "    for pattern in gapped_patterns:\n",
    "        prefix = pattern[ : k - 1] + pattern[k    : (k * 2)]\n",
    "        suffix = pattern[1: k + 1] + pattern[k + 2: (k * 2) + 1]\n",
    "        if prefix in graph:\n",
    "            graph[prefix].append(suffix)\n",
    "        else:\n",
    "            graph[prefix] = [suffix]\n",
    "\n",
    "    return(graph)\n",
    "\n",
    "def PairedReconstruction(gapped_patterns, k, d):\n",
    "    graph = PairedPathGraph(gapped_patterns, k)\n",
    "    path  = EulerianPath(graph)  \n",
    "    \n",
    "    gapped_patterns_1 = []\n",
    "    gapped_patterns_2 = []\n",
    "    \n",
    "    for gapped_pattern in path:\n",
    "        pattern_1, pattern_2 = gapped_pattern.split('|')\n",
    "        gapped_patterns_1.append(pattern_1)\n",
    "        gapped_patterns_2.append(pattern_2)\n",
    "        \n",
    "    prefix_string = NaiveLink(gapped_patterns_1)\n",
    "    suffix_string = NaiveLink(gapped_patterns_2)\n",
    "    \n",
    "    overlap_length = len(prefix_string) - k - d\n",
    "    for i in range(overlap_length):\n",
    "        if suffix_string[i] != prefix_string[i + k + d]:\n",
    "            return('there is no string spelled by the gapped patterns')\n",
    "    whole_string = prefix_string + suffix_string[overlap_length:]\n",
    "    return(whole_string)\n",
    "\n",
    "# Test\n",
    "k = 4\n",
    "d = 2\n",
    "gapped_patterns = '''GAGA|TTGA\n",
    "TCGT|GATG\n",
    "CGTG|ATGT\n",
    "TGGT|TGAG\n",
    "GTGA|TGTT\n",
    "GTGG|GTGA\n",
    "TGAG|GTTG\n",
    "GGTC|GAGA\n",
    "GTCG|AGAT'''\n",
    "\n",
    "PairedReconstruction(gapped_patterns, k, d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['TGT', 'TGGA', 'GAT', 'ATG', 'ATG', 'CAT', 'AGA']"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Contig Generation Problem: Generate the contigs from a collection of reads (with imperfect coverage).\n",
    "Input: A collection of k-mers Patterns. \n",
    "Output: All contigs in DeBruijn(Patterns).\n",
    "'''\n",
    "\n",
    "def CountInOut(graph):\n",
    "    if type(graph) == str:\n",
    "        graph = GraphString2Dict(graph)\n",
    "        \n",
    "    in_dict  = {}\n",
    "    out_dict = {}\n",
    "    \n",
    "    for key, values in graph.items():\n",
    "        out_dict[key] = 0\n",
    "        for value in values:\n",
    "            in_dict[value] = 0\n",
    "    \n",
    "    for prefix, suffixs in graph.items():\n",
    "        out_dict[prefix] = out_dict[prefix] + len(suffixs)\n",
    "        for suffix in suffixs:\n",
    "            in_dict[suffix] = in_dict[suffix] + 1\n",
    "            \n",
    "    return (in_dict,out_dict)\n",
    "\n",
    "def MaximalNonBranchingPaths(strings):\n",
    "    graph = PathGraph(strings)\n",
    "\n",
    "    in_count, out_count = CountInOut(graph)\n",
    "\n",
    "    nodes = set()\n",
    "    for _ in in_count.keys():\n",
    "        nodes.add(_)\n",
    "    for _ in out_count.keys():    \n",
    "        nodes.add(_)\n",
    "    nodes = list(nodes)\n",
    "\n",
    "    '''\n",
    "    #This is for isolated cycle\n",
    "    one_one_graph = {}\n",
    "    for key, value in graph.items():\n",
    "        if (in_count.get(key, -1) == out_count.get(key, -1) == 1):\n",
    "            one_one_graph[key] = value\n",
    "    '''\n",
    "    \n",
    "    paths = []\n",
    "    for node in nodes:\n",
    "        if (in_count.get(node) != 1) | (out_count.get(node) != 1):\n",
    "            if out_count.get(node, -1) > 0:\n",
    "                for out_node in graph[node]:\n",
    "                    non_branching_path = []\n",
    "                    non_branching_path.append(node)\n",
    "                    non_branching_path.append(out_node)\n",
    "                    while (in_count.get(out_node, -1) == out_count.get(out_node, -1) == 1):\n",
    "                        #one_one_graph.pop(out_node)\n",
    "                        non_branching_path.extend(graph[out_node])\n",
    "                        out_node = graph[out_node][0]\n",
    "                    paths.append(NaiveLink(non_branching_path))\n",
    "                    \n",
    "    return(paths)\n",
    "\n",
    "# Test\n",
    "strings = '''ATG\n",
    "ATG\n",
    "TGT\n",
    "TGG\n",
    "CAT\n",
    "GGA\n",
    "GAT\n",
    "AGA'''\n",
    "\n",
    "MaximalNonBranchingPaths(strings)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MAMAPRTEINSTRING'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Protein Translation Problem: Translate an RNA string into an amino acid string.\n",
    "Input: An RNA string Pattern and the array GeneticCode.\n",
    "Output: The translation of Pattern into an amino acid string Peptide.\n",
    "'''\n",
    "\n",
    "\n",
    "table = open('data/RNA_codon_table.txt').read()\n",
    "table = table.split('\\n')\n",
    "translation_rule = {}\n",
    "for line in table:\n",
    "    RNA, peptide = line.split(' ')\n",
    "    translation_rule[RNA] = peptide\n",
    "\n",
    "def RNA2peptide(RNA_string):\n",
    "    peptide_string = ''\n",
    "    position       = 0\n",
    "    while ('*' not in peptide_string) & (position in range(len(RNA_string)-2)):\n",
    "        RNA            = RNA_string[position: position + 3]\n",
    "        AA             = translation_rule[RNA]\n",
    "        peptide_string = peptide_string + AA\n",
    "        position       = position + 3\n",
    "        \n",
    "    peptide_string = peptide_string[: -1]\n",
    "    \n",
    "    return(peptide_string)\n",
    "\n",
    "# Test\n",
    "RNA_string = 'AUGGCCAUGGCGCCCAGAACUGAGAUCAAUAGUACCCGUAUUAACGGGUGA'\n",
    "\n",
    "RNA2peptide(RNA_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GGCCAT', 'ATGGCC', 'ATGGCC']"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Peptide Encoding Problem: Find substrings of a genome encoding a given amino acid sequence.\n",
    "Input: A DNA string Text, an amino acid string Peptide, and the array GeneticCode.\n",
    "Output: All substrings of Text encoding Peptide (if any such substrings exist).\n",
    "'''\n",
    "\n",
    "def RNA2peptideNoStop(RNA_string):\n",
    "    peptide_string = ''\n",
    "    position       = 0\n",
    "    while (position in range(len(RNA_string) - 2)):\n",
    "        RNA            = RNA_string[position: position + 3]\n",
    "        AA             = translation_rule[RNA]\n",
    "        peptide_string = peptide_string + AA\n",
    "        position       = position + 3\n",
    "    \n",
    "    return(peptide_string)\n",
    "\n",
    "def ReverseComplement(DNA_string):\n",
    "    complementry_rule = {\n",
    "        'A' : 'T',\n",
    "        'T' : 'A',\n",
    "        'C' : 'G',\n",
    "        'G' : 'C' \n",
    "    }\n",
    "    com_string = ''\n",
    "    for base in DNA_string:\n",
    "        com_string = com_string + complementry_rule[base]\n",
    "    \n",
    "    com_string = com_string[::-1]\n",
    "    \n",
    "    return (com_string)\n",
    "\n",
    "def DNA2RNA(DNA_string):\n",
    "    complementry_rule = {\n",
    "        'A' : 'U',\n",
    "        'T' : 'A',\n",
    "        'C' : 'G',\n",
    "        'G' : 'C' \n",
    "    }\n",
    "    RNA_string = ''\n",
    "    for base in DNA_string:\n",
    "        RNA_string = RNA_string + complementry_rule[base]\n",
    "    \n",
    "    RNA_string = RNA_string[::-1]\n",
    "    \n",
    "    return(RNA_string)\n",
    "\n",
    "def RNA2DNA(RNA_string):\n",
    "    complementry_rule = {\n",
    "        'A' : 'T',\n",
    "        'U' : 'A',\n",
    "        'C' : 'G',\n",
    "        'G' : 'C' \n",
    "    }\n",
    "    DNA_string = ''\n",
    "    for base in RNA_string:\n",
    "        DNA_string = DNA_string + complementry_rule[base]\n",
    "    \n",
    "    DNA_string = DNA_string[::-1]\n",
    "    \n",
    "    return(DNA_string)\n",
    "\n",
    "def PatternMatching(pattern, genome):\n",
    "    positions = []\n",
    "    for start in range(len(genome) - len(pattern) + 1):\n",
    "        end = start + len(pattern)\n",
    "        if pattern == genome[start: end]:\n",
    "            positions.append(start)\n",
    "    return(positions)\n",
    "\n",
    "def PeptideEncoding(DNA_string, peptide_pattern):\n",
    "    matched_DNAs   = []\n",
    "    for start in range(3):\n",
    "        DNA_pattern    = DNA_string\n",
    "        RNA_string     = DNA2RNA(DNA_pattern)[start:]\n",
    "        peptide_string = RNA2peptideNoStop(RNA_string)\n",
    "        positions      = PatternMatching(peptide_pattern, peptide_string)\n",
    "        \n",
    "        for position in positions:\n",
    "            start       = position * 3\n",
    "            end         = start + len(peptide_pattern) * 3\n",
    "            matched_RNA = RNA_string[start: end]\n",
    "            matched_DNAs.append(RNA2DNA(matched_RNA))\n",
    "            \n",
    "    for start in range(3):\n",
    "        DNA_pattern    = ReverseComplement(DNA_string)\n",
    "        RNA_string     = DNA2RNA(DNA_pattern)[start:]\n",
    "        peptide_string = RNA2peptideNoStop(RNA_string)\n",
    "        positions      = PatternMatching(peptide_pattern, peptide_string)\n",
    "        for position in positions:\n",
    "            start       = position * 3\n",
    "            end         = start + len(peptide_pattern) * 3\n",
    "            matched_RNA = RNA_string[start: end]\n",
    "            matched_DNAs.append(ReverseComplement(RNA2DNA(matched_RNA)))\n",
    "    \n",
    "    return(matched_DNAs)\n",
    "\n",
    "# Test\n",
    "DNA_string = 'ATGGCCATGGCCCCCAGAACTGAGATCAATAGTACCCGTATTAACGGGTGA'\n",
    "peptide_pattern = 'MA'\n",
    "\n",
    "PeptideEncoding(DNA_string, peptide_pattern)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0, 113, 114, 128, 129, 242, 242, 257, 370, 371, 484]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Code Challenge: Implement LinearSpectrum.\n",
    "Input: An amino acid string Peptide.\n",
    "Output: The linear spectrum of Peptide.\n",
    "'''\n",
    "\n",
    "table = open('data/integer_mass_table.txt').read()\n",
    "table = table.split('\\n')\n",
    "AA_weitght = {}\n",
    "for line in table:\n",
    "    AA, weight = line.split(' ')\n",
    "    AA_weitght[AA] = int(weight)\n",
    "\n",
    "def LinearSpectrum(peptide_string):\n",
    "    prefix_weight = [0]\n",
    "    for AA in peptide_string:\n",
    "        prefix_weight.append(prefix_weight[-1] + AA_weitght[AA])\n",
    "        \n",
    "    linear_spectrum = [0]\n",
    "    for i in range(len(peptide_string)):\n",
    "        for j in range(i + 1, len(peptide_string) + 1):\n",
    "            linear_spectrum.append(prefix_weight[j] - prefix_weight[i])\n",
    "    linear_spectrum.sort()\n",
    "    return(linear_spectrum)\n",
    "\n",
    "# Test\n",
    "peptide_string = 'NQEL'\n",
    "\n",
    "LinearSpectrum(peptide_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0, 113, 114, 128, 129, 227, 242, 242, 257, 355, 356, 370, 371, 484]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Generating Theoretical Spectrum Problem: Generate the theoretical spectrum of a cyclic peptide.\n",
    "Input: An amino acid string Peptide.\n",
    "Output: Cyclospectrum(Peptide).\n",
    "'''\n",
    "\n",
    "table = open('data/integer_mass_table.txt').read()\n",
    "table = table.split('\\n')\n",
    "AA_weitght = {}\n",
    "for line in table:\n",
    "    AA, weight = line.split(' ')\n",
    "    AA_weitght[AA] = int(weight)\n",
    "\n",
    "def CyclicSpectrum(peptide_string):\n",
    "    prefix_weight = [0]\n",
    "    for AA in peptide_string:\n",
    "        prefix_weight.append(prefix_weight[-1] + AA_weitght[AA])\n",
    "    \n",
    "    total_weight = prefix_weight[-1]\n",
    "    cyclic_spectrum = [0]\n",
    "    for i in range(len(peptide_string)):\n",
    "        for j in range(i + 1, len(peptide_string) + 1):\n",
    "            cyclic_spectrum.append(prefix_weight[j] - prefix_weight[i])\n",
    "            if (i > 0) and (j < (len(peptide_string))):\n",
    "                cyclic_spectrum.append(total_weight - (prefix_weight[j] - prefix_weight[i]))\n",
    "    cyclic_spectrum.sort()\n",
    "    return(cyclic_spectrum)\n",
    "\n",
    "# Test\n",
    "peptide_string = 'LEQN'\n",
    "\n",
    "CyclicSpectrum(peptide_string)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['IKW', 'IQW', 'IWK', 'IWQ', 'LKW', 'LQW', 'LWK', 'LWQ', 'KIW', 'KLW', 'KWI', 'KWL', 'QIW', 'QLW', 'QWI', 'QWL', 'WIK', 'WIQ', 'WLK', 'WLQ', 'WKI', 'WKL', 'WQI', 'WQL']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['186-113-128',\n",
       " '186-128-113',\n",
       " '113-128-186',\n",
       " '113-186-128',\n",
       " '128-186-113',\n",
       " '128-113-186']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Code Challenge: Implement CyclopeptideSequencing\n",
    "'''\n",
    "\n",
    "def CyclopeptideSequencing(spectrum):\n",
    "    if type(spectrum) == str:\n",
    "        spectrum = spectrum.split(' ')\n",
    "    spectrum = list(map(int, spectrum))\n",
    "    AAs = ['G','A','S','P','V','T','C','I','L','N','D','K','Q','E','M','H','F','R','Y','W']\n",
    "    \n",
    "    output_peptides = []\n",
    "    \n",
    "    peptides = ['']\n",
    "    while len(peptides) != 0:\n",
    "        \n",
    "        extend_peptides = []\n",
    "        for peptide in peptides:\n",
    "            for AA in AAs:\n",
    "                extend_peptide = peptide + AA\n",
    "                extend_peptides.append(extend_peptide)\n",
    "        peptides = copy.deepcopy(extend_peptides)\n",
    "        \n",
    "        for peptide in peptides:\n",
    "            peptide_spectrum = LinearSpectrum(peptide)\n",
    "            if peptide_spectrum[-1] == spectrum[-1]:\n",
    "                if CyclicSpectrum(peptide) == spectrum:\n",
    "                    output_peptides.append(peptide)\n",
    "                extend_peptides.remove(peptide)\n",
    "            \n",
    "            else:\n",
    "                for weight in peptide_spectrum:\n",
    "                    if weight not in spectrum:\n",
    "                        extend_peptides.remove(peptide)\n",
    "                        break\n",
    "                        \n",
    "        peptides = copy.deepcopy(extend_peptides)\n",
    "    print(output_peptides)   \n",
    "    \n",
    "    output_weight_string = []\n",
    "    for output_peptide in output_peptides:\n",
    "        weight_string = []\n",
    "        for AA in output_peptide:\n",
    "            weight = str(AA_weitght[AA])\n",
    "            weight_string.append(weight)\n",
    "        output_weight_string.append('-'.join(weight_string))\n",
    "        \n",
    "    output_weight_string = list(set(output_weight_string))\n",
    "    \n",
    "    return(output_weight_string)\n",
    "\n",
    "# Test\n",
    "spectrum = '0 113 128 186 241 299 314 427'\n",
    "\n",
    "CyclopeptideSequencing(spectrum)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Cyclopeptide Scoring Problem: Compute the score of a cyclic peptide against a spectrum.\n",
    "Input: An amino acid string Peptide and a collection of integers Spectrum. \n",
    "Output: The score of Peptide against Spectrum, Score(Peptide, Spectrum).\n",
    "'''\n",
    "\n",
    "def Cyclopeptide_Scoring(peptide_string, spectrum):\n",
    "    if type(spectrum) == str:\n",
    "        spectrum = spectrum.split(' ')\n",
    "    spectrum = list(map(int, spectrum))\n",
    "    \n",
    "    theoretical_spectrum = CyclicSpectrum(peptide_string)\n",
    "    \n",
    "    score = 0\n",
    "    for mass in theoretical_spectrum:\n",
    "        if mass in spectrum:\n",
    "            spectrum.remove(mass)\n",
    "            score = score + 1\n",
    "            \n",
    "    return(score)\n",
    "\n",
    "# Test\n",
    "peptide_string = 'NQEL'\n",
    "spectrum = '0 99 113 114 128 227 257 299 355 356 370 371 484'\n",
    "\n",
    "Cyclopeptide_Scoring(peptide_string, spectrum)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['LAST', 'ALST']"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Code Challenge: Implement Trim (reproduced below).\n",
    "Input: A collection of peptides Leaderboard, a collection of integers Spectrum, and an integer N.\n",
    "Output: The N highest-scoring linear peptides on Leaderboard with respect to Spectrum.\n",
    "'''\n",
    "\n",
    "def Linearpeptide_Scoring(peptide_string, spectrum):\n",
    "    if type(spectrum) == str:\n",
    "        spectrum = spectrum.split(' ')\n",
    "    spectrum = list(map(int, spectrum))\n",
    "    \n",
    "    theoretical_spectrum = LinearSpectrum(peptide_string)\n",
    "    \n",
    "    score = 0\n",
    "    for mass in theoretical_spectrum:\n",
    "        if mass in spectrum:\n",
    "            spectrum.remove(mass)\n",
    "            score = score + 1\n",
    "            \n",
    "    return(score)\n",
    "\n",
    "def Trim(leaderboard, spectrum, N):\n",
    "    if type(spectrum) == str:\n",
    "        spectrum = spectrum.split(' ')\n",
    "    spectrum = list(map(int, spectrum))\n",
    "    \n",
    "    if type(leaderboard) == str:\n",
    "        leaderboard = leaderboard.split(' ')\n",
    "    \n",
    "    linear_scores = []\n",
    "    for peptide in leaderboard:\n",
    "        \n",
    "        peptide_score = Linearpeptide_Scoring(peptide, spectrum)\n",
    "        linear_scores.append(peptide_score)\n",
    "    \n",
    "    score_peptide      = zip(linear_scores, leaderboard)\n",
    "    score_peptide      = sorted(score_peptide, reverse = True)\n",
    "    score_sorted       = []\n",
    "    leaderboard_sorted = []\n",
    "    for x, y in score_peptide:\n",
    "        score_sorted.append(x)\n",
    "        leaderboard_sorted.append(y)\n",
    "        \n",
    "    for rank in range(N, len(leaderboard)):\n",
    "        if score_sorted[rank] < score_sorted[N - 1]:\n",
    "            leaderboard_sorted = leaderboard_sorted[: rank]\n",
    "            return(leaderboard_sorted)\n",
    "        \n",
    "    return(leaderboard_sorted)\n",
    "\n",
    "# Test\n",
    "leaderboard = 'LAST ALST TLLT TQAS'\n",
    "spectrum = '0 71 87 101 113 158 184 188 259 271 372'\n",
    "N = 2\n",
    "\n",
    "Trim(leaderboard, spectrum, N)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'113-147-71-129'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Code Challenge: Implement LeaderboardCyclopeptideSequencing.\n",
    "Input: An integer N and a collection of integers Spectrum.\n",
    "Output: LeaderPeptide after running LeaderboardCyclopeptideSequencing(Spectrum, N).\n",
    "'''\n",
    "\n",
    "def LeaderboardCyclopeptideSequencing(spectrum, N):\n",
    "    if type(spectrum) == str:\n",
    "        spectrum = spectrum.split(' ')\n",
    "    spectrum = list(map(int, spectrum))\n",
    "    AAs = ['G','A','S','P','V','T','C','I','L','N','D','K','Q','E','M','H','F','R','Y','W']\n",
    "    \n",
    "    leaderboard   = ['']\n",
    "    leaderpeptide = ''\n",
    "    while len(leaderboard) != 0:\n",
    "        extend_leaderboard = []\n",
    "        for peptide in leaderboard:\n",
    "            for AA in AAs:\n",
    "                extend_peptide = peptide + AA\n",
    "                extend_leaderboard.append(extend_peptide)\n",
    "        leaderboard = copy.deepcopy(extend_leaderboard)\n",
    "        \n",
    "        for peptide in leaderboard:\n",
    "            peptide_spectrum = LinearSpectrum(peptide)\n",
    "            if peptide_spectrum[-1] == spectrum[-1]:\n",
    "                if Linearpeptide_Scoring(peptide, spectrum) > Linearpeptide_Scoring(leaderpeptide, spectrum):\n",
    "                    leaderpeptide = peptide\n",
    "            elif peptide_spectrum[-1] > spectrum[-1]:\n",
    "                extend_leaderboard.remove(peptide)\n",
    "\n",
    "        leaderboard = Trim(extend_leaderboard, spectrum, N)\n",
    "            \n",
    "    output_number = []\n",
    "    for AA in leaderpeptide:\n",
    "        weight = str(AA_weitght[AA])\n",
    "        output_number.append(weight)\n",
    "    output_number = '-'.join(output_number)\n",
    "    \n",
    "    return(output_number)      \n",
    "\n",
    "# Test\n",
    "N = 10\n",
    "spectrum = '0 71 113 129 147 200 218 260 313 331 347 389 460'\n",
    "\n",
    "LeaderboardCyclopeptideSequencing(spectrum, N)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[137, 186, 186, 137]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Spectral Convolution Problem: Compute the convolution of a spectrum.\n",
    "Input: A collection of integers Spectrum.\n",
    "Output: The list of elements in the convolution of Spectrum. If an element has multiplicity k, it should appear exactly k times.\n",
    "'''\n",
    "\n",
    "def SpectralConvolution(spectrum):\n",
    "    if type(spectrum) == str:\n",
    "        spectrum = spectrum.split(' ')\n",
    "    spectrum = list(map(int, spectrum))\n",
    "    \n",
    "    convolution_spectrum = []\n",
    "    for i in spectrum:\n",
    "        for j in spectrum:\n",
    "            if ((i - j) > 56) & ((i - j) < 200):\n",
    "                convolution_spectrum.append(i - j)\n",
    "                \n",
    "    #convolution_spectrum.sort()\n",
    "    return(convolution_spectrum)\n",
    "\n",
    "spectrum = '0 137 186 323'\n",
    "\n",
    "SpectralConvolution(spectrum)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[71, 99, 129, 57, 79, 58]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Code Challenge: Implement ConvolutionCyclopeptideSequencing.\n",
    "Input: An integer M, an integer N, and a collection of (possibly repeated) integers Spectrum.\n",
    "Output: A cyclic peptide LeaderPeptide with amino acids taken only from the top M elements (and ties) of the convolution of Spectrum that fall between 57 and 200, and where the size of Leaderboard is restricted to the top N (and ties).\n",
    "'''\n",
    "\n",
    "def LinearSpectrum_Scoring(weights, spectrum):\n",
    "    \n",
    "    theoretical_spectrum = LinearWeightSpectrum(weights)\n",
    "    copy_spectrum        = copy.deepcopy(spectrum)\n",
    "    \n",
    "    score = 0\n",
    "    for mass in theoretical_spectrum:\n",
    "        if mass in copy_spectrum:\n",
    "            copy_spectrum.remove(mass)\n",
    "            score = score + 1\n",
    "    return(score)\n",
    "\n",
    "def LinearWeightSpectrum(weights):\n",
    "    if type(weights) == str:\n",
    "        weights = weights.split('-')\n",
    "    weights = list(map(int, weights))\n",
    "    \n",
    "    prefix_weight = [0]\n",
    "    for weight in weights:\n",
    "        prefix_weight.append(prefix_weight[-1] + weight)\n",
    "\n",
    "    linear_spectrum = [0]\n",
    "    for i in range(len(weights)):\n",
    "        for j in range(i + 1, len(weights) + 1):\n",
    "            linear_spectrum.append(prefix_weight[j] - prefix_weight[i])\n",
    "    linear_spectrum.sort()\n",
    "    \n",
    "    return(linear_spectrum)\n",
    "\n",
    "def Trim_spectrum(leaderboard, spectrum, N):\n",
    "    if type(spectrum) == str:\n",
    "        spectrum = spectrum.split(' ')\n",
    "    spectrum = list(map(int, spectrum))\n",
    "    \n",
    "    if type(leaderboard) == str:\n",
    "        leaderboard = leaderboard.split(' ')\n",
    "    \n",
    "    linear_scores = []\n",
    "    for weight_spectrum in leaderboard:    \n",
    "        spectrum_score = LinearSpectrum_Scoring(weight_spectrum, spectrum)\n",
    "        linear_scores.append(spectrum_score)\n",
    "    \n",
    "    score_specturm     = zip(linear_scores, leaderboard)\n",
    "    score_specturm     = sorted(score_specturm, reverse = True)\n",
    "    score_sorted       = []\n",
    "    leaderboard_sorted = []\n",
    "    for x, y in score_specturm:\n",
    "        score_sorted.append(x)\n",
    "        leaderboard_sorted.append(y)\n",
    "        \n",
    "    for rank in range(N, len(leaderboard)):\n",
    "        if score_sorted[rank] < score_sorted[N - 1]:\n",
    "            leaderboard_sorted = leaderboard_sorted[: rank]\n",
    "            return(leaderboard_sorted)\n",
    "        \n",
    "    return(leaderboard_sorted)\n",
    "\n",
    "def ConvolutionCyclopeptideSequencing(M, N, spectrum):\n",
    "\n",
    "    if type(spectrum) == str:\n",
    "        spectrum = spectrum.split(' ')\n",
    "    spectrum = list(map(int, spectrum))\n",
    "    spectrum.sort()\n",
    "    \n",
    "    # Pick top M weight\n",
    "    spectral_convolution = SpectralConvolution(spectrum)\n",
    "    convolution_count    = Counter(spectral_convolution)\n",
    "    \n",
    "    weights = list(convolution_count.keys())\n",
    "    counts  = list(convolution_count.values())\n",
    "    \n",
    "    counts_weights = zip(counts,weights)\n",
    "    counts_weights = sorted(counts_weights, reverse = True)\n",
    "    count_sorted   = []\n",
    "    weight_sorted  = []\n",
    "    for x, y in counts_weights:\n",
    "        count_sorted.append(x)\n",
    "        weight_sorted.append(y)\n",
    "    \n",
    "    for rank in range(M, len(count_sorted)):\n",
    "        if count_sorted[rank] < count_sorted[M - 1]:\n",
    "            weight_sorted = weight_sorted[: rank]\n",
    "    \n",
    "    weight_candidates = weight_sorted\n",
    "\n",
    "    # Inital leaderboard with weight candidates\n",
    "    leaderboard   = []\n",
    "    for weight_candidate in weight_candidates:\n",
    "        leaderboard.append([weight_candidate])\n",
    "    leader_weight = [0]\n",
    "    \n",
    "    while len(leaderboard) != 0:\n",
    "        extend_leaderboard = []\n",
    "        for weight_string in leaderboard:\n",
    "            for weight in weight_candidates:\n",
    "                extend_weight = copy.deepcopy(weight_string)\n",
    "                extend_weight.append(weight)\n",
    "                extend_leaderboard.append(extend_weight)\n",
    "                \n",
    "        leaderboard = copy.deepcopy(extend_leaderboard)\n",
    "\n",
    "        for weights in leaderboard:\n",
    "            weights_spectrum = LinearWeightSpectrum(weights)\n",
    "            if weights_spectrum[-1] == spectrum[-1]:\n",
    "                if LinearSpectrum_Scoring(weights, spectrum) > LinearSpectrum_Scoring(leader_weight, spectrum):\n",
    "                    leader_weight = weights\n",
    "                    \n",
    "            elif weights_spectrum[-1] > spectrum[-1]:\n",
    "                extend_leaderboard.remove(weights)\n",
    "        \n",
    "        # Select top N weights for next loop leaderboard\n",
    "        leaderboard = Trim_spectrum(extend_leaderboard, spectrum, N)\n",
    "        \n",
    "    return(leader_weight)\n",
    "\n",
    "M = 20\n",
    "N = 60\n",
    "spectrum = '57 57 71 99 129 137 170 186 194 208 228 265 285 299 307 323 356 364 394 422 493'\n",
    "\n",
    "ConvolutionCyclopeptideSequencing(M, N, spectrum)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
