{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import copy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 1: Introduction to Sequence Alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "966"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Code Challenge: Solve the Change Problem. The DPChange pseudocode is reproduced below for your convenience.\n",
    "Input: An integer money and an array Coins = (coin1, ..., coind).\n",
    "Output: The minimum number of coins with denominations Coins that changes money.\n",
    "'''\n",
    "\n",
    "def DPChange(money, coins):\n",
    "    if type(coins) == str:\n",
    "        coins = coins.split(',')\n",
    "    coins = list(map(int, coins))\n",
    "    \n",
    "    min_count    = [float('Inf')] * (money + 1)\n",
    "    min_count[0] = 0\n",
    "    \n",
    "    for m in range(1, (money + 1)):\n",
    "        for coin in coins:\n",
    "            if m >= coin:\n",
    "                if min_count[m - coin] + 1 < min_count[m]:\n",
    "                    min_count[m] = min_count[m - coin] + 1\n",
    "    return(min_count[money])\n",
    "    \n",
    "# Test\n",
    "money = 17373\n",
    "coins = '18,5,3,1'\n",
    "\n",
    "DPChange(money, coins)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "34"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Code Challenge: Find the length of a longest path in the Manhattan Tourist Problem.\n",
    "Input: Integers n and m, followed by an n × (m + 1) matrix Down and an (n + 1) × m matrix Right.\n",
    "Output: The length of a longest path from source (0, 0) to sink (n, m) in the n × m rectangular grid whose edges are defined by the matrices Down and Right.\n",
    "'''\n",
    "\n",
    "def ManhattanTourist(n, m, down, right):\n",
    "    # reformat down and right matrix\n",
    "    if type(down) == str:\n",
    "        down = down.split('\\n')\n",
    "        for i in range(len(down)):\n",
    "            down[i] = down[i].split(' ')\n",
    "        down = np.array(down).reshape(n, m + 1)\n",
    "    down = down.astype(int)\n",
    "    \n",
    "    if type(right) == str:\n",
    "        right = right.split('\\n')\n",
    "        for i in range(len(right)):\n",
    "            right[i] = right[i].split(' ')\n",
    "        right = np.array(right).reshape(n + 1, m)\n",
    "    right = right.astype(int)\n",
    "    \n",
    "    # initalize matrix\n",
    "    matrix = np.zeros((n + 1, m + 1))\n",
    "    for i in range(1, n + 1):\n",
    "        matrix[i, 0] = matrix[i - 1, 0] + down[i - 1, 0]\n",
    "    for j in range(1, m + 1):\n",
    "        matrix[0, j] = matrix[0, j - 1] + right[0, j - 1]\n",
    "    \n",
    "    # fill matrix\n",
    "    for i in range(1, n + 1):\n",
    "        for j in range(1, m + 1):\n",
    "            matrix[i, j] = max(matrix[i - 1, j] + down[i - 1, j],\n",
    "                               matrix[i, j - 1] + right[i, j - 1])\n",
    "            \n",
    "    return(int(matrix[n, m]))\n",
    "\n",
    "# Test\n",
    "n = 4\n",
    "m = 4\n",
    "down = '''1 0 2 4 3\n",
    "4 6 5 2 1\n",
    "4 4 5 2 1\n",
    "5 6 8 5 3'''\n",
    "\n",
    "right = '''3 2 4 0\n",
    "3 2 4 2\n",
    "0 7 3 3\n",
    "3 3 0 2\n",
    "1 3 2 2'''\n",
    "\n",
    "ManhattanTourist(n, m, down, right)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AACTTG'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\"\"\"\n",
    "Code Challenge: Use OutputLCS (reproduced below) to solve the Longest Common Subsequence Problem.\n",
    "Input: Two strings s and t.\n",
    "Output: A longest common subsequence of s and t.\n",
    "\"\"\"\n",
    "\n",
    "def LCSBackTrack(v, w):\n",
    "    \n",
    "    matrix = np.zeros((len(v) + 1, len(w) + 1))\n",
    "    backtrack = np.zeros((len(v) + 1, len(w) + 1), '<U32')\n",
    "    for i in range(1, (len(v) + 1)):\n",
    "        for j in range(1, (len(w) + 1)):\n",
    "            if v[i - 1] == w[j - 1]:\n",
    "                 match = 1\n",
    "            else:\n",
    "                 match = 0\n",
    "            matrix[i, j] = max(matrix[i - 1, j], matrix[i, j - 1], matrix[ i - 1, j - 1] + match)\n",
    "            \n",
    "            if matrix[i, j] == matrix[i - 1, j]:\n",
    "                backtrack[i, j] = 'deletion'\n",
    "            elif matrix[i, j] == matrix[i, j - 1]:\n",
    "                backtrack[i, j] = 'insertion'\n",
    "            elif (matrix[i, j] == matrix[i - 1, j - 1] + 1) & (v[i - 1] == w[j - 1]):\n",
    "                backtrack[i, j] = 'diagonal'\n",
    "    return(backtrack)\n",
    "\n",
    "def OutputLCS(backtrack, v, i, j, longest_string):\n",
    "    if (i == 0)|(j == 0):\n",
    "        return\n",
    "    if backtrack[i, j] == 'deletion':\n",
    "        OutputLCS(backtrack, v, i - 1, j, longest_string)\n",
    "    elif backtrack[i, j] == 'insertion':\n",
    "        OutputLCS(backtrack, v, i, j - 1, longest_string)\n",
    "    else:\n",
    "        OutputLCS(backtrack, v, i - 1, j - 1, longest_string)\n",
    "        longest_string.append(v[i - 1])\n",
    "\n",
    "def LongestCommonSubsequence(v, w):\n",
    "    backtrack = LCSBackTrack(v, w)\n",
    "    i = len(v) \n",
    "    j = len(w) \n",
    "    longest_string = []\n",
    "    \n",
    "    OutputLCS(backtrack, v, i, j, longest_string)\n",
    "    \n",
    "    longest_string = ''.join(longest_string)\n",
    "    \n",
    "    return(longest_string)\n",
    "\n",
    "# Test\n",
    "v = 'AACCTTGG'\n",
    "w = 'ACACTGTGA'\n",
    "\n",
    "LongestCommonSubsequence(v, w)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(9, '0->2->3->4')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Code Challenge: Solve the Longest Path in a DAG Problem.\n",
    "Input: An integer representing the starting node to consider in a graph, followed by an integer representing the ending node to \n",
    "    consider, followed by a list of edges in the graph. The edge notation \"0->1:7\" indicates that an edge connects node 0 to node 1\n",
    "    with weight 7.  You may assume a given topological order corresponding to nodes in increasing order.\n",
    "Output: The length of a longest path in the graph, followed by a longest path. \n",
    "'''\n",
    "\n",
    "def GraphString2Dict(graph):\n",
    "    graph = graph.split('\\n')\n",
    "    graph_dict   = {}\n",
    "    graph_weight = {}\n",
    "    \n",
    "    for path in graph:\n",
    "        from_, to_    = path.split('->')\n",
    "        to_, weight = to_.split(':')\n",
    "        graph_weight[''.join([from_,'->',to_])] = int(weight)\n",
    "        if from_ in graph_dict:\n",
    "            graph_dict[from_].append(to_)\n",
    "        else:\n",
    "            graph_dict[from_]=[to_]\n",
    "    \n",
    "    return(graph_dict,graph_weight)\n",
    "\n",
    "def Longest_Path(start_point, end_point, graph):\n",
    "    start_point, end_point = str(start_point), str(end_point)\n",
    "    graph, graph_weight    = GraphString2Dict(graph)\n",
    "    \n",
    "    longest_path = []\n",
    "    score_list   = [0] * 100\n",
    "    track_back   = ['000'] * 100\n",
    "    froms        = [start_point]\n",
    "\n",
    "    while len(froms) != 0:\n",
    "        next_from = []\n",
    "        for from_ in froms:\n",
    "            tos = graph[from_]\n",
    "            for to in tos:\n",
    "                weight_key = from_ + '->' + to\n",
    "                score_list[int(to)] = max(score_list[int(to)], score_list[int(from_)] + graph_weight[weight_key])\n",
    "                if score_list[int(to)] == (score_list[int(from_)] + graph_weight[weight_key]):\n",
    "                    track_back[int(to)] = from_\n",
    "                if to in graph:\n",
    "                    next_from.append(to)\n",
    "            froms = next_from\n",
    "    \n",
    "    to = end_point\n",
    "    longest_path.append(end_point)\n",
    "    \n",
    "    while to != start_point:\n",
    "        from_ = track_back[int(to)]\n",
    "        longest_path.append(from_)\n",
    "        to    = from_\n",
    "        \n",
    "    longest_path = '->'.join(longest_path[::-1])\n",
    "    \n",
    "    return(score_list[int(end_point)],longest_path)\n",
    "\n",
    "# Test\n",
    "start_point = 0\n",
    "end_point   = 4\n",
    "graph = '''0->1:7\n",
    "0->2:4\n",
    "2->3:2\n",
    "1->4:1\n",
    "3->4:3'''\n",
    "\n",
    "Longest_Path(start_point, end_point, graph)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 2: From Finding a Longest Path to Aligning DNA Strings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8, 'PLEASANTLY', '-MEA--N-LY')"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Code Challenge: Solve the Global Alignment Problem.\n",
    "Input: Two protein strings written in the single-letter amino acid alphabet.\n",
    "Output: The maximum alignment score of these strings followed by an alignment achieving this maximum score. Use the BLOSUM62 scoring matrix for matches and mismatches as well as the indel penalty σ = 5.\n",
    "'''\n",
    "\n",
    "BLOSUM62 = pd.read_csv('data/BLOSUM62.csv')\n",
    "BLOSUM62 = BLOSUM62.set_index('X')\n",
    "\n",
    "def BackTrack_ScoreMatrix_Global(v, w, score_matrix, indel_pentalty):\n",
    "    \n",
    "    matrix = np.full((len(v) + 1, len(w) + 1), 0, 'int')\n",
    "    backtrack = np.zeros((len(v) + 1, len(w) + 1), '<U32')\n",
    "    \n",
    "    for i in range(1, (len(v) + 1)):\n",
    "        matrix[i, 0] = matrix[i - 1, 0] - indel_pentalty\n",
    "    for j in range(1, (len(w) + 1)):\n",
    "        matrix[0, j] = matrix[0, j - 1] - indel_pentalty\n",
    "        \n",
    "    for i in range(1, (len(v) + 1)):\n",
    "        for j in range(1, (len(w) + 1)):\n",
    "            \n",
    "            matrix[i, j] = max(matrix[i - 1, j]     - indel_pentalty, \n",
    "                               matrix[i    , j - 1] - indel_pentalty, \n",
    "                               matrix[i - 1, j - 1] + score_matrix.at[v[i - 1], w[j - 1]])\n",
    "            \n",
    "            if matrix[i, j] == matrix[i - 1, j] - indel_pentalty:\n",
    "                backtrack[i, j] = 'deletion'\n",
    "                \n",
    "            elif matrix[i, j] == matrix[i, j - 1] - indel_pentalty:\n",
    "                backtrack[i, j] = 'insertion'\n",
    "                \n",
    "            elif matrix[i, j] == (matrix[ i - 1, j - 1]) + (score_matrix.at[v[i - 1], w[j - 1]]):\n",
    "                backtrack[i, j] = 'diagonal'\n",
    "            \n",
    "    return(matrix[len(v), len(w)],backtrack)\n",
    "\n",
    "def AppendAlignment(backtrack, v, w, i, j, aligned_v, aligned_w):\n",
    "    if (i == 0)|(j == 0):\n",
    "        if (i == 0)&(j == 0):\n",
    "            return\n",
    "        \n",
    "        elif backtrack[i, j] == 'deletion':\n",
    "            if i == 0:\n",
    "                aligned_v.append('-')\n",
    "            else : \n",
    "                aligned_v.append(v[i - 1])\n",
    "            aligned_w.append('-')\n",
    "            \n",
    "        elif backtrack[i, j] == 'insertion':\n",
    "            if j == 0:\n",
    "                aligned_w.append('-')\n",
    "            else :\n",
    "                aligned_w.append(w[j - 1])\n",
    "            aligend_v.append('-')         \n",
    "\n",
    "        else:\n",
    "            if i == 0:\n",
    "                aligned_v.append('-')\n",
    "            else : \n",
    "                aligned_v.append(v[i - 1])\n",
    "            if j == 0:\n",
    "                aligned_w.append('-')\n",
    "            else :\n",
    "                aligned_w.append(w[j - 1])   \n",
    "        return\n",
    "    \n",
    "    if backtrack[i, j] == 'jump':\n",
    "        return\n",
    "        \n",
    "    elif backtrack[i, j] == 'deletion':\n",
    "        AppendAlignment(backtrack, v, w, i - 1, j, aligned_v, aligned_w)\n",
    "        aligned_w.append('-')\n",
    "        aligned_v.append(v[i - 1])\n",
    "        \n",
    "    elif backtrack[i, j] == 'insertion':\n",
    "        AppendAlignment(backtrack, v, w, i, j - 1, aligned_v, aligned_w)\n",
    "        aligned_v.append('-')\n",
    "        aligned_w.append(w[j - 1])\n",
    "        \n",
    "    else:\n",
    "        AppendAlignment(backtrack, v, w, i - 1, j - 1, aligned_v, aligned_w)\n",
    "        aligned_v.append(v[i - 1])\n",
    "        aligned_w.append(w[j - 1])\n",
    "\n",
    "def Score_Alignment_Global(v, w, score_matrix, indel_pentalty):\n",
    "    score, backtrack = BackTrack_ScoreMatrix_Global(v, w, score_matrix, indel_pentalty)\n",
    "\n",
    "    i = len(v) \n",
    "    j = len(w) \n",
    "    aligned_v = []\n",
    "    aligned_w = []\n",
    "\n",
    "    AppendAlignment(backtrack, v, w, i, j, aligned_v, aligned_w)\n",
    "    aligned_v = ''.join(aligned_v)\n",
    "    aligned_w = ''.join(aligned_w)\n",
    "    \n",
    "    return(score, aligned_v, aligned_w)\n",
    "\n",
    "# Test\n",
    "v = 'PLEASANTLY'\n",
    "w = 'MEANLY'\n",
    "score_matrix = BLOSUM62\n",
    "indel_pentalty = 5\n",
    "\n",
    "Score_Alignment_Global(v, w, score_matrix, indel_pentalty)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(15, 'EANL-Y', 'ENALTY')"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Code Challenge: Solve the Local Alignment Problem.\n",
    "Input: Two protein strings written in the single-letter amino acid alphabet.\n",
    "Output: The maximum score of a local alignment of the strings, followed by a local alignment of these strings achieving the maximum\n",
    "     score. Use the PAM250 scoring matrix for matches and mismatches as well as the indel penalty σ = 5.\n",
    "'''\n",
    "\n",
    "PAM250 = pd.read_csv('data/PAM250.csv')\n",
    "PAM250 = PAM250.set_index('X')\n",
    "\n",
    "def BackTrack_ScoreMatrix_Local(v, w, score_matrix):\n",
    "    \n",
    "    matrix = np.full((len(v) + 1, len(w) + 1), 0, 'int')\n",
    "    backtrack = np.zeros((len(v) + 1, len(w) + 1), '<U32')\n",
    "    \n",
    "    for i in range(1, (len(v) + 1)):\n",
    "        matrix[i, 0] = matrix[i - 1, 0] - 5\n",
    "    for j in range(1, (len(w) + 1)):\n",
    "        matrix[0, j] = matrix[0, j - 1] - 5\n",
    "        \n",
    "    for i in range(1, (len(v) + 1)):\n",
    "        for j in range(1, (len(w) + 1)):\n",
    "            \n",
    "            matrix[i, j] = max(0,\n",
    "                               matrix[i - 1, j]     - 5, \n",
    "                               matrix[i    , j - 1] - 5, \n",
    "                               matrix[i - 1, j - 1] + score_matrix.at[v[i - 1], w[j - 1]])\n",
    "            \n",
    "            if matrix[i, j] == 0:\n",
    "                backtrack[i, j] = 'jump'\n",
    "                \n",
    "            elif matrix[i, j] == matrix[i - 1, j] - 5:\n",
    "                backtrack[i, j] = 'deletion'\n",
    "                \n",
    "            elif matrix[i, j] == matrix[i, j - 1] - 5:\n",
    "                backtrack[i, j] = 'insertion'\n",
    "                \n",
    "            elif matrix[i, j] == (matrix[ i - 1, j - 1]) + (score_matrix.at[v[i - 1], w[j - 1]]):\n",
    "                backtrack[i, j] = 'diagonal'\n",
    "            \n",
    "    i_start, j_start = np.where(matrix == np.max(matrix))\n",
    "    \n",
    "    return(np.max(matrix), backtrack, int(i_start), int(j_start))\n",
    "\n",
    "def Score_Alignment_Local(v, w, score_matrix):\n",
    "    score, backtrack, i, j = BackTrack_ScoreMatrix_Local(v, w, score_matrix)\n",
    "    \n",
    "    #i = len(v) \n",
    "    #j = len(w) \n",
    "    aligned_v = []\n",
    "    aligned_w = []\n",
    "\n",
    "    AppendAlignment(backtrack, v, w, i, j, aligned_v, aligned_w)\n",
    "    aligned_v = ''.join(aligned_v)\n",
    "    aligned_w = ''.join(aligned_w)\n",
    "    \n",
    "    return(score, aligned_v, aligned_w)\n",
    "\n",
    "# Test\n",
    "v = 'MEANLY'\n",
    "w = 'PENALTY'\n",
    "score_matrix = PAM250\n",
    "\n",
    "Score_Alignment_Local(v, w, score_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Edit Distance Problem: Find the edit distance between two strings.\n",
    "Input: Two strings.\n",
    "Output: The edit distance between these strings.\n",
    "'''\n",
    "\n",
    "def HammingDistance(string1, string2):\n",
    "    mismatch = 0\n",
    "    for base1, base2 in zip (string1, string2):\n",
    "        if base1 != base2: \n",
    "            mismatch = mismatch + 1\n",
    "    return (mismatch)\n",
    "\n",
    "def EditDistance(v, w):\n",
    "    score_matrix         = pd.DataFrame(np.identity(20)) - 1\n",
    "    score_matrix.columns = PAM250.columns\n",
    "    score_matrix.index   = PAM250.index\n",
    "    indel_pentalty       = 1\n",
    "    \n",
    "    score, aligned_v, aligned_w = Score_Alignment_Global(v, w, score_matrix, indel_pentalty)\n",
    "\n",
    "    edit_distance = HammingDistance(aligned_v, aligned_w)\n",
    "    \n",
    "    return(edit_distance)\n",
    "\n",
    "# Test\n",
    "v = 'PLEASANTLY'\n",
    "w = 'MEANLY'\n",
    "\n",
    "EditDistance(v, w)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2, 'TAGGCTTA', 'TAGA-T-A')"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Code Challenge: Solve the Fitting Alignment Problem.\n",
    "Input: Two nucleotide strings v and w, where v has length at most 1000 and w has length at most 100.\n",
    "Output: A highest-scoring fitting alignment between v and w. Use the simple scoring method in which matches count +1 and both the\n",
    "     mismatch and indel penalties are 1.\n",
    "'''\n",
    "\n",
    "def BackTrack_ScoreMatrix_Fitting(v, w, score_matrix, indel_pentalty):\n",
    "    \n",
    "    matrix = np.full((len(v) + 1, len(w) + 1), 0, 'int')\n",
    "    backtrack = np.zeros((len(v) + 1, len(w) + 1), '<U32')\n",
    "    \n",
    "    '''for i in range(1, (len(v) + 1)):\n",
    "        matrix[i, 0] = matrix[i - 1, 0] - indel_pentalty\n",
    "    for j in range(1, (len(w) + 1)):\n",
    "        matrix[0, j] = matrix[0, j - 1] - indel_pentalty'''\n",
    "        \n",
    "    for i in range(1, (len(v) + 1)):\n",
    "        for j in range(1, (len(w) + 1)):\n",
    "            \n",
    "            matrix[i, j] = max(\n",
    "                               matrix[i - 1, j]     - indel_pentalty, \n",
    "                               matrix[i    , j - 1] - indel_pentalty, \n",
    "                               matrix[i - 1, j - 1] + (score_matrix.at[v[i - 1],w[j - 1]]))\n",
    "            \n",
    "            \n",
    "                \n",
    "            if matrix[i, j] == matrix[i - 1, j] - indel_pentalty:\n",
    "                backtrack[i, j] = 'deletion'\n",
    "                \n",
    "            elif matrix[i, j] == matrix[i, j - 1] - indel_pentalty:\n",
    "                backtrack[i, j] = 'insertion'\n",
    "                \n",
    "            elif matrix[i, j] == (matrix[ i - 1, j - 1]) + (score_matrix.at[v[i - 1], w[j - 1]]):\n",
    "                backtrack[i, j] = 'diagonal'\n",
    "                \n",
    "    return(matrix, backtrack )\n",
    "\n",
    "def AppendAlignment_Fitting(backtrack, v, w, i, j, aligned_v, aligned_w):\n",
    "\n",
    "    if (i == 0)|(j == 0):\n",
    "        return\n",
    "        \n",
    "    if backtrack[i, j] == 'deletion':\n",
    "        AppendAlignment(backtrack, v, w, i - 1, j, aligned_v, aligned_w)\n",
    "        aligned_w.append('-')\n",
    "        aligned_v.append(v[i - 1])\n",
    "        \n",
    "    elif backtrack[i, j] == 'insertion':\n",
    "        AppendAlignment(backtrack, v, w, i, j - 1, aligned_v, aligned_w)\n",
    "        aligned_v.append('-')\n",
    "        aligned_w.append(w[j - 1])\n",
    "        \n",
    "    else:\n",
    "        AppendAlignment(backtrack, v, w, i - 1, j - 1, aligned_v, aligned_w)\n",
    "        aligned_v.append(v[i - 1])\n",
    "        aligned_w.append(w[j - 1])\n",
    "\n",
    "def Score_Alignment_Fitting(v, w, score_matrix, indel_pentalty):\n",
    "    \n",
    "    matrix, backtrack = BackTrack_ScoreMatrix_Fitting(v, w, score_matrix, indel_pentalty)\n",
    "\n",
    "    j = len(w) \n",
    "    i = np.where(matrix[j : len(v), len(w)] == np.max(matrix[j : len(v), len(w)]))[0][0] + j\n",
    "\n",
    "    score = np.max(matrix[:,j])\n",
    "\n",
    "    aligned_v = []\n",
    "    aligned_w = []\n",
    "\n",
    "    AppendAlignment_Fitting(backtrack, v, w, i, j, aligned_v, aligned_w)\n",
    "    \n",
    "    aligned_v = ''.join(aligned_v)\n",
    "    aligned_w = ''.join(aligned_w)\n",
    "    \n",
    "    while aligned_v[0] != aligned_w[0]:\n",
    "        aligned_v = aligned_v[1:]\n",
    "        aligned_w = aligned_w[1:]\n",
    "    \n",
    "    return(score, aligned_v, aligned_w )\n",
    "\n",
    "# Test\n",
    "score_matrix         = (pd.DataFrame(np.identity(20)) * 2) - 1\n",
    "score_matrix.columns = PAM250.columns\n",
    "score_matrix.index   = PAM250.index\n",
    "v = 'GTAGGCTTAAGGTTA'\n",
    "w = 'TAGATA'\n",
    "indel_pentalty = 1\n",
    "\n",
    "Score_Alignment_Fitting(v, w, score_matrix, indel_pentalty)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1, 'HEAE', 'HEAG')"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Code Challenge: Solve the Overlap Alignment Problem.\n",
    "Input: Two strings v and w, each of length at most 1000.\n",
    "Output: The score of an optimal overlap alignment of v and w, followed by an alignment of a suffix v' of v and a prefix w' of w\n",
    "     achieving this maximum score. Use an alignment score in which matches count +1 and both the mismatch and indel penalties are 2.\n",
    "'''\n",
    "\n",
    "def BackTrack_ScoreMatrix_Overlap(v, w, score_matrix, indel_pentalty):\n",
    "    \n",
    "    matrix = np.full((len(v) + 1, len(w) + 1), 0, 'int')\n",
    "    backtrack = np.zeros((len(v) + 1, len(w) + 1), '<U32')\n",
    "    \n",
    "    '''for i in range(1, (len(v) + 1)):\n",
    "        matrix[i, 0] = matrix[i - 1, 0] - indel_pentalty'''\n",
    "    '''for j in range(1, (len(w) + 1)):\n",
    "        matrix[0, j] = matrix[0, j - 1] - indel_pentalty'''\n",
    "        \n",
    "    for i in range(1, (len(v) + 1)):\n",
    "        for j in range(1, (len(w) + 1)):\n",
    "            if (j == 1) | (i == len(v)):\n",
    "                matrix[i, j] = max(0,\n",
    "                               matrix[i - 1, j]     - indel_pentalty, \n",
    "                               matrix[i    , j - 1] - indel_pentalty, \n",
    "                               matrix[i - 1, j - 1] + (score_matrix.at[v[i - 1],w[j - 1]]))\n",
    "            else:\n",
    "                matrix[i, j] = max(\n",
    "                               matrix[i - 1, j]     - indel_pentalty, \n",
    "                               matrix[i    , j - 1] - indel_pentalty, \n",
    "                               matrix[i - 1, j - 1] + (score_matrix.at[v[i - 1],w[j - 1]]))\n",
    "                \n",
    "            if matrix[i, j] == matrix[i - 1, j] - indel_pentalty:\n",
    "                backtrack[i, j] = 'deletion'\n",
    "                \n",
    "            elif matrix[i, j] == matrix[i, j - 1] - indel_pentalty:\n",
    "                backtrack[i, j] = 'insertion'\n",
    "                \n",
    "            elif matrix[i, j] == (matrix[ i - 1, j - 1]) + (score_matrix.at[v[i - 1], w[j - 1]]):\n",
    "                backtrack[i, j] = 'diagonal'\n",
    "                \n",
    "    return(matrix, backtrack )\n",
    "\n",
    "def Score_Alignment_Overlap(v, w, score_matrix, indel_pentalty):\n",
    "    \n",
    "    matrix, backtrack = BackTrack_ScoreMatrix_Overlap(v, w, score_matrix, indel_pentalty)\n",
    "\n",
    "    i = len(v) \n",
    "    j = np.where(matrix[i, :] == np.max(matrix[i, :]))[0][-1]\n",
    "    score = matrix[i, j]\n",
    "\n",
    "    aligned_v = []\n",
    "    aligned_w = []\n",
    "\n",
    "    AppendAlignment(backtrack, v, w, i, j, aligned_v, aligned_w)\n",
    "    aligned_v = ''.join(aligned_v)\n",
    "    aligned_w = ''.join(aligned_w)\n",
    "\n",
    "    while aligned_v[0] != aligned_w[0]:\n",
    "        aligned_v = aligned_v[1:]\n",
    "        aligned_w = aligned_w[1:]\n",
    "\n",
    "    return(score,aligned_v,aligned_w)\n",
    "\n",
    "# Test\n",
    "score_matrix         = (pd.DataFrame(np.identity(20)) * 3) - 2\n",
    "score_matrix.columns = PAM250.columns\n",
    "score_matrix.index   = PAM250.index\n",
    "\n",
    "v = 'PAWHEAE'\n",
    "w = 'HEAGAWGHEE'\n",
    "indel_pentalty = 2\n",
    "\n",
    "Score_Alignment_Overlap(v, w, score_matrix, indel_pentalty)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 3: Advanced Topics in Sequence Alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8, 'PRT---EINS', 'PRTWPSEIN-')"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Code Challenge: Solve the Alignment with Affine Gap Penalties Problem.\n",
    "Input: Two amino acid strings v and w (each of length at most 100).\n",
    "Output: The maximum alignment score between v and w, followed by an alignment of v and w achieving this maximum score. Use the\n",
    "     BLOSUM62 scoring matrix, a gap opening penalty of 11, and a gap extension penalty of 1.\n",
    "'''\n",
    "\n",
    "def Global_Alignment_with_Affine_Gap(v, w, score_matrix, opening_penalty, extension_penalty):\n",
    "    \n",
    "    # Initalize backtrack, matrix with penalty.\n",
    "    matrix_upper  = np.full((len(v) + 1, len(w) + 1), 0, 'float')\n",
    "    matrix_middle = np.full((len(v) + 1, len(w) + 1), 0, 'float')\n",
    "    matrix_lower  = np.full((len(v) + 1, len(w) + 1), 0, 'float')\n",
    "\n",
    "    backtrack_upper  = np.full((len(v) + 1, len(w) + 1), 0, '<U32')\n",
    "    backtrack_middle = np.full((len(v) + 1, len(w) + 1), 0, '<U32')\n",
    "    backtrack_lower  = np.full((len(v) + 1, len(w) + 1), 0, '<U32')\n",
    "\n",
    "    matrix_middle[1, 0] = matrix_middle[0, 1] = matrix_upper[0, 1] = matrix_lower[1, 0] = -1 * opening_penalty\n",
    "    matrix_upper[:, 0]  = matrix_lower[0, :]  = -1 * float('Inf')\n",
    "\n",
    "    for i in range(2, (len(v) + 1)):\n",
    "        matrix_middle[i, 0] = matrix_middle[i - 1, 0] - extension_penalty\n",
    "        matrix_lower[i, 0]  = matrix_lower[i - 1, 0]  - extension_penalty\n",
    "\n",
    "    for j in range(2, (len(w) + 1)):\n",
    "        matrix_middle[0, j] = matrix_middle[0, j - 1] - extension_penalty\n",
    "        matrix_upper[0, j]  = matrix_upper[0, j - 1]  - extension_penalty\n",
    "\n",
    "    # fill the score\n",
    "    for i in range(1, len(v)+1):\n",
    "        for j in range(1, len(w)+1):\n",
    "            matrix_lower[i, j] = max(matrix_lower[i - 1, j] - extension_penalty,\n",
    "                                    matrix_middle[i - 1, j] - opening_penalty)\n",
    "\n",
    "            matrix_upper[i, j] = max(matrix_upper[i, j - 1] - extension_penalty,\n",
    "                                    matrix_middle[i, j - 1] - opening_penalty)\n",
    "\n",
    "            matrix_middle[i, j] = max(matrix_lower[i, j],\n",
    "                                      matrix_middle[i - 1, j - 1] + score_matrix.at[v[i-1], w[j-1]],\n",
    "                                      matrix_upper[i, j])\n",
    "\n",
    "            if matrix_lower[i, j] == (matrix_lower[i - 1, j] - extension_penalty):\n",
    "                backtrack_lower[i, j] = 'down'\n",
    "            else: \n",
    "                backtrack_lower[i, j] = 'lower_2_middle'\n",
    "\n",
    "            if matrix_upper[i, j] == (matrix_upper[i, j - 1] - extension_penalty):\n",
    "                backtrack_upper[i, j] = 'right'\n",
    "            else:\n",
    "                backtrack_upper[i, j] = 'upper_2_middle'\n",
    "\n",
    "            if matrix_middle[i, j] == matrix_lower[i, j]:\n",
    "                backtrack_middle[i, j] = 'lower'\n",
    "            elif matrix_middle[i, j] == matrix_upper[i, j]:\n",
    "                backtrack_middle[i, j] = 'upper'\n",
    "            else:\n",
    "                backtrack_middle[i, j] = 'diagonal'\n",
    "    \n",
    "    i = len(v) \n",
    "    j = len(w)\n",
    "    score = max(matrix_lower[i, j],\n",
    "               matrix_middle[i, j],\n",
    "                matrix_upper[i, j])\n",
    "    \n",
    "    if score == matrix_lower[i, j]:\n",
    "        current_matrix = backtrack_lower\n",
    "        \n",
    "    elif score == matrix_upper[i, j]:\n",
    "        current_matrix = backtrack_upper\n",
    "        \n",
    "    else:\n",
    "        current_matrix = backtrack_middle\n",
    "    \n",
    "    #backtrack\n",
    "    aligned_v = []\n",
    "    aligned_w = []\n",
    "    \n",
    "    while (i * j) != 0:\n",
    "        if current_matrix[i, j] == 'lower_2_middle':\n",
    "            aligned_v.append(v[i - 1])\n",
    "            aligned_w.append('-')\n",
    "            i = i - 1\n",
    "            current_matrix = backtrack_middle\n",
    "            \n",
    "        elif current_matrix[i, j] == 'down':\n",
    "            aligned_v.append(v[i - 1])\n",
    "            aligned_w.append('-')\n",
    "            i = i - 1\n",
    "        \n",
    "        elif current_matrix[i, j] == 'upper_2_middle':\n",
    "            aligned_v.append('-')\n",
    "            aligned_w.append(w[j - 1])\n",
    "            j = j - 1\n",
    "            current_matrix = backtrack_middle\n",
    "            \n",
    "        elif current_matrix[i, j] == 'right':\n",
    "            aligned_v.append('-')\n",
    "            aligned_w.append(w[j - 1])\n",
    "            j = j - 1\n",
    "        \n",
    "        elif current_matrix[i, j] == 'lower':\n",
    "            current_matrix = backtrack_lower\n",
    "            \n",
    "        elif current_matrix[i, j] == 'upper':\n",
    "            current_matrix = backtrack_upper\n",
    "        \n",
    "        elif current_matrix[i, j] == 'diagonal':\n",
    "            aligned_v.append(v[i - 1])\n",
    "            aligned_w.append(w[j - 1])\n",
    "            i = i - 1\n",
    "            j = j - 1\n",
    "    \n",
    "    aligned_v = ''.join(aligned_v[::-1])\n",
    "    aligned_w = ''.join(aligned_w[::-1])\n",
    "    \n",
    "    return(int(score), aligned_v, aligned_w)\n",
    "\n",
    "# Test\n",
    "v = 'PRTEINS'\n",
    "w = 'PRTWPSEIN'\n",
    "score_matrix      = BLOSUM62\n",
    "opening_penalty   = 11\n",
    "extension_penalty = 1\n",
    "\n",
    "Global_Alignment_with_Affine_Gap(v, w, score_matrix, opening_penalty, extension_penalty)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[4, 3] [5, 4]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "('diagonal', 4)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Code Challenge: Solve the Middle Edge in Linear Space Problem (for protein strings).\n",
    "Input: Two amino acid strings.\n",
    "Output: A middle edge in the alignment graph in the form \"(i, j) (k, l)\", where (i, j) connects to (k, l). To compute scores, use the\n",
    "     BLOSUM62 scoring matrix and a (linear) indel penalty equal to 5.\n",
    "'''\n",
    "\n",
    "def Middle_Edge(v, w, score_matrix, indel_pentalty):\n",
    "    middle_col = len(w) // 2\n",
    "    matrix     = np.full((len(v) + 1, middle_col + 2), 0, 'float')\n",
    "    \n",
    "    for i in range(1, (len(v) + 1)):\n",
    "        matrix[i, 0] = matrix[i - 1, 0] - indel_pentalty\n",
    "    for j in range(1, (middle_col + 2)):\n",
    "        matrix[0, j] = matrix[0, middle_col - 1] - indel_pentalty\n",
    "    \n",
    "    for i in range(1, (len(v) + 1)):\n",
    "        for j in range(1, (middle_col + 2)):\n",
    "            \n",
    "            matrix[i, j] = max(matrix[i - 1, j]     - indel_pentalty, \n",
    "                               matrix[i    , j - 1] - indel_pentalty, \n",
    "                               matrix[i - 1, j - 1] + (score_matrix.at[v[i - 1],w[j - 1]]))\n",
    "            \n",
    "    j = middle_col        \n",
    "    i = int(np.where(matrix[:, j] == np.max(matrix[:, j]))[0])\n",
    "\n",
    "    if max(matrix[i + 1, j + 1], matrix[i + 1, j], matrix[i, j + 1]) == matrix[i + 1, j + 1]:\n",
    "        direction = 'diagonal'\n",
    "        k, l      = i + 1, j + 1\n",
    "        \n",
    "    elif max(matrix[i + 1, j + 1], matrix[i + 1, j], matrix[i, j + 1]) == matrix[i, j + 1]:\n",
    "        direction = 'right'\n",
    "        k, l      = i, j + 1\n",
    "    else:\n",
    "        direction = 'bottom'\n",
    "        k, l      = i + 1, j\n",
    "\n",
    "    print([i, j], [k, l])\n",
    "    return(direction, i)\n",
    "\n",
    "# Test\n",
    "v = 'PLEASANTLY'\n",
    "w = 'MEASNLY'\n",
    "score_matrix   = BLOSUM62\n",
    "indel_pentalty = 5\n",
    "\n",
    "Middle_Edge(v, w, score_matrix, indel_pentalty)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3, 'AT---ATC--CG-', '-T-----C-C-GA', 'ATGTA--CT--G-')"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Code Challenge: Solve the Multiple Longest Common Subsequence Problem.\n",
    "Input: Three DNA strings of length at most 10.\n",
    "Output: The length of a longest common subsequence of these three strings, followed by a multiple alignment of the three strings\n",
    "     corresponding to such an alignment.\n",
    "'''\n",
    "\n",
    "\n",
    "def Three_String_Global_Alignment(s1, s2, s3 ,match_score, unmatch_score):\n",
    "    matrix    = np.full((len(s1) + 1, len(s2) + 1, len(s3) + 1), 0, 'float')\n",
    "    backtrack = np.zeros((len(s1) + 1, len(s2) + 1, len(s3) + 1), dtype='int')\n",
    "    \n",
    "    # fill the matrix\n",
    "    for i in range(1, (len(s1) + 1)):\n",
    "        for j in range(1, (len(s2) + 1)):\n",
    "            for k in range(1, (len(s3) + 1)):\n",
    "                \n",
    "                if s1[i -1] == s2[j - 1] == s3[k - 1]:\n",
    "                    match = match_score\n",
    "                else:\n",
    "                    match = unmatch_score\n",
    "                    \n",
    "                scores = [matrix[i - 1, j    , k    ] + unmatch_score,\n",
    "                          matrix[i    , j - 1, k    ] + unmatch_score,\n",
    "                          matrix[i    , j    , k - 1] + unmatch_score,\n",
    "                          matrix[i - 1, j - 1, k    ] + unmatch_score,\n",
    "                          matrix[i    , j - 1, k - 1] + unmatch_score,\n",
    "                          matrix[i - 1, j    , k - 1] + unmatch_score,\n",
    "                          matrix[i - 1, j - 1, k - 1] + match]\n",
    "                \n",
    "                \n",
    "                matrix[i, j, k]    = max(scores)\n",
    "                backtrack[i, j, k] = scores.index(matrix[i, j, k])\n",
    "\n",
    "    aligned_s1 = []\n",
    "    aligned_s2 = []\n",
    "    aligned_s3 = []\n",
    "    \n",
    "    i, j, k = len(s1), len(s2), len(s3)\n",
    "    \n",
    "    # trackback\n",
    "    while (i * j * k) != 0 :\n",
    "        \n",
    "        if backtrack[i, j, k] == 0:\n",
    "            aligned_s1.append(s1[i - 1])\n",
    "            aligned_s2.append('-')\n",
    "            aligned_s3.append('-') \n",
    "            i = i - 1\n",
    "        \n",
    "        elif backtrack[i, j, k] == 1:\n",
    "            aligned_s1.append('-')\n",
    "            aligned_s2.append(s2[j - 1])\n",
    "            aligned_s3.append('-') \n",
    "            j = j - 1\n",
    "        \n",
    "        elif backtrack[i, j, k] == 2:\n",
    "            aligned_s1.append('-')\n",
    "            aligned_s2.append('-')\n",
    "            aligned_s3.append(s3[k - 1]) \n",
    "            k = k - 1\n",
    "            \n",
    "        elif backtrack[i, j, k] == 3:\n",
    "            aligned_s1.append(s1[i - 1])\n",
    "            aligned_s2.append(s2[j - 1])\n",
    "            aligned_s3.append('-') \n",
    "            i, j = i - 1, j - 1\n",
    "        \n",
    "        elif backtrack[i, j, k] == 4:\n",
    "            aligned_s1.append('-')\n",
    "            aligned_s2.append(s2[j - 1])\n",
    "            aligned_s3.append(s3[k - 1]) \n",
    "            j, k = j - 1, k - 1\n",
    "        \n",
    "        elif backtrack[i, j, k] == 5:\n",
    "            aligned_s1.append(s1[i - 1])\n",
    "            aligned_s2.append('-')\n",
    "            aligned_s3.append(s3[k - 1]) \n",
    "            i, k = i - 1, k - 1\n",
    "        \n",
    "        elif backtrack[i, j, k] == 6:\n",
    "            aligned_s1.append(s1[i - 1])\n",
    "            aligned_s2.append(s2[j - 1])\n",
    "            aligned_s3.append(s3[k - 1]) \n",
    "            i, j, k = i - 1, j - 1, k - 1\n",
    "    \n",
    "    # fill the rest of backtrack\n",
    "    while (i + j + k) != 0:\n",
    "        if i > 0:\n",
    "            aligned_s1.append(s1[i - 1])\n",
    "            i = i - 1\n",
    "        else:\n",
    "            aligned_s1.append('-')\n",
    "        \n",
    "        if j > 0:\n",
    "            aligned_s2.append(s2[j - 1])\n",
    "            j = j - 1\n",
    "        else:\n",
    "            aligned_s2.append('-')\n",
    "        \n",
    "        if k > 0:\n",
    "            aligned_s3.append(s3[k - 1])\n",
    "            k = k - 1\n",
    "        else:\n",
    "            aligned_s3.append('-')\n",
    "            \n",
    "    \n",
    "    aligned_s1 = ''.join(aligned_s1[::-1])\n",
    "    aligned_s2 = ''.join(aligned_s2[::-1])\n",
    "    aligned_s3 = ''.join(aligned_s3[::-1])\n",
    "                            \n",
    "    output_score = int(matrix[len(s1), len(s2), len(s3)] )    \n",
    "    \n",
    "    return(output_score, aligned_s1, aligned_s2, aligned_s3)\n",
    "\n",
    "# Test\n",
    "s1 = 'ATATCCG'\n",
    "s2 = 'TCCGA'\n",
    "s3 = 'ATGTACTG'\n",
    "match_score   = 1\n",
    "unmatch_score = 0\n",
    "\n",
    "Three_String_Global_Alignment(s1, s2, s3 ,match_score, unmatch_score)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 4: Genome Rearrangements and Fragility"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(-3 +4 +1 +5 -2)\n",
      "(-1 -4 +3 +5 -2)\n",
      "(+1 -4 +3 +5 -2)\n",
      "(+1 +2 -5 -3 +4)\n",
      "(+1 +2 +3 +5 +4)\n",
      "(+1 +2 +3 -4 -5)\n",
      "(+1 +2 +3 +4 -5)\n",
      "(+1 +2 +3 +4 +5)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "7"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Code Challenge: Implement GreedySorting.\n",
    "Input: A permutation P.\n",
    "Output: The sequence of permutations corresponding to applying GreedySorting to P, ending with the identity permutation.\n",
    "'''\n",
    "\n",
    "def reverse(p_list):\n",
    "    for i in range(len(p_list)):\n",
    "        p_list[i] = p_list[i] * -1\n",
    "    p_list = p_list[::-1]\n",
    "    return(p_list)\n",
    "\n",
    "def print_p(p_list):\n",
    "    p_list = list(map(str, p_list))\n",
    "    p_list = p_list[1:]\n",
    "    for i in range(len(p_list)):\n",
    "        if (p_list[i][0]) != '-':\n",
    "            p_list[i] = '+' + p_list[i]\n",
    "    p_list = ' '.join(p_list)\n",
    "    print('(' + p_list + ')')\n",
    "\n",
    "def GreedySorting(p):\n",
    "    if type(p) == str:\n",
    "        p = p.split(' ')\n",
    "    p = list(map(int, p))\n",
    "    \n",
    "    approxReversalDistance = 0\n",
    "    p = [0] + p\n",
    "    \n",
    "    for k in range(1,len(p)):\n",
    "        while p[k] != k:\n",
    "            print_p(p)\n",
    "            approxReversalDistance = approxReversalDistance + 1\n",
    "            \n",
    "            if p[k] == -k:\n",
    "                p[k] = p[k] * -1\n",
    "\n",
    "            elif (-k) in p:\n",
    "                k_index = p.index(-k)\n",
    "                p = p[: k] + reverse(p[k : k_index + 1]) + p[k_index + 1 :]\n",
    "\n",
    "            else:\n",
    "                k_index = p.index(k)\n",
    "                p = p[: k] + reverse(p[k : k_index + 1]) + p[k_index + 1 :]\n",
    "    \n",
    "    print_p(p)  \n",
    "    return(approxReversalDistance)\n",
    "    \n",
    "# Test\n",
    "p = '-3 +4 +1 +5 -2'\n",
    "\n",
    "GreedySorting(p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Number of Breakpoints Problem: Find the number of breakpoints in a permutation.\n",
    "Input: A permutation.\n",
    "Output: The number of breakpoints in this permutation.\n",
    "'''\n",
    "\n",
    "def Count_Breakpoint(permutation):\n",
    "    if type(permutation) == str:\n",
    "        permutation = permutation.split(' ')\n",
    "    permutation = list(map(int, permutation))\n",
    "    \n",
    "    breakpoint_count = 0\n",
    "    \n",
    "    permutation = [0] + permutation + [len(permutation) + 1]\n",
    "    \n",
    "    for i in range(len(permutation) - 1):\n",
    "        if permutation[i] != permutation[i + 1] - 1:\n",
    "            breakpoint_count = breakpoint_count + 1\n",
    "    \n",
    "    return(breakpoint_count)\n",
    "\n",
    "# Test\n",
    "permutation = '+3 +4 +5 -12 -8 -7 -6 +1 +2 +10 +9 -11 +13 +14'\n",
    "\n",
    "Count_Breakpoint(permutation)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Week 5: Applying Genome Rearrangement Analysis to Find Genome Fragility"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1, 2, 4, 3, 6, 5, 7, 8]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Code Challenge: Implement ChromosomeToCycle.\n",
    "Input: A chromosome Chromosome containing n synteny blocks.\n",
    "Output: The sequence Nodes of integers between 1 and 2n resulting from applying ChromosomeToCycle to Chromosome.\n",
    "'''\n",
    "\n",
    "def ChromosomeToCycle(chromosome):\n",
    "    if type(chromosome) == str:\n",
    "        chromosome = chromosome.split(' ')\n",
    "    chromosome = list(map(int, chromosome))\n",
    "    \n",
    "    nodes = []\n",
    "    for old_node in chromosome:\n",
    "        \n",
    "        if old_node > 0:\n",
    "            nodes.extend([2 * old_node - 1, 2 * old_node])\n",
    "            \n",
    "        else :\n",
    "            nodes.extend([-2 * old_node   , -2 * old_node - 1])\n",
    "            \n",
    "    return(nodes)\n",
    "\n",
    "# Test\n",
    "chromosome = '+1 -2 -3 +4'\n",
    "\n",
    "ChromosomeToCycle(chromosome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1, -2, -3, 4]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Code Challenge: Implement CycleToChromosome.\n",
    "Input: A sequence Nodes of integers between 1 and 2n.\n",
    "Output: The chromosome Chromosome containing n synteny blocks resulting from applying CycleToChromosome to Nodes.\n",
    "'''\n",
    "\n",
    "def CycleToChromosome(nodes):\n",
    "    if type(nodes) == str:\n",
    "        nodes = nodes.split(' ')\n",
    "    nodes = list(map(int, nodes))\n",
    "    nodes = [0] + nodes\n",
    "    \n",
    "    chromosome = []\n",
    "    \n",
    "    for i in range(1, int(len(nodes) / 2) + 1):\n",
    "        \n",
    "        if nodes[2 * i - 1] < nodes[2 * i]:\n",
    "            node = nodes[2 * i] / 2\n",
    "        \n",
    "        else:\n",
    "            node = -1 * nodes[(2 * i) - 1] / 2\n",
    "\n",
    "        chromosome.append(int(node))\n",
    "        \n",
    "    return(chromosome)\n",
    "\n",
    "# Test\n",
    "nodes = '1 2 4 3 6 5 7 8'\n",
    "\n",
    "CycleToChromosome(nodes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[2, 4], [3, 6], [5, 1], [8, 9], [10, 12], [11, 7]]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Code Challenge: Implement ColoredEdges.\n",
    "Input: A genome P.\n",
    "Output: The collection of colored edges in the genome graph of P in the form (x, y).\n",
    "'''\n",
    "\n",
    "def ColoredEdges(p):\n",
    "    if type(p) == str:\n",
    "        p = p[1 : -1]\n",
    "        p = p.split(')(')\n",
    "    \n",
    "    edges = []\n",
    "    \n",
    "    for chromosome in p:\n",
    "        if type(chromosome) == str:\n",
    "            chromosome = chromosome.split(' ')\n",
    "            chromosome = list(map(int, chromosome))\n",
    "            \n",
    "        nodes      = ChromosomeToCycle(chromosome)\n",
    "        for i in range(1, int(len(chromosome))):\n",
    "            edge = [nodes[2 * i - 1], nodes[2 * i]]\n",
    "            edges.append(edge)\n",
    "        last_egde = [nodes[-1], nodes[0]]\n",
    "        edges.append(last_egde)\n",
    "        \n",
    "    return(edges)\n",
    "\n",
    "# Test\n",
    "p = '(+1 -2 -3)(+4 +5 -6)'\n",
    "\n",
    "ColoredEdges(p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[1, -2, -3], [-4, 5, -6]]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Code Challenge: Implement GraphToGenome.\n",
    "Input: The colored edges ColoredEdges of a genome graph.\n",
    "Output: The genome P corresponding to this genome graph.\n",
    "'''\n",
    "\n",
    "def GraphToGenome(genome_graph):\n",
    "    \n",
    "    if type(genome_graph) == str:\n",
    "        genome_graph = genome_graph.replace('(','')\n",
    "        genome_graph = genome_graph.replace(')','')\n",
    "        genome_graph = genome_graph.replace(' ','')\n",
    "        genome_graph = genome_graph.split(',')\n",
    "        genome_graph = list(map(int, genome_graph))\n",
    "    \n",
    "    else:\n",
    "        tmp_genome_graph = []\n",
    "        for sub_genome_graph in genome_graph:\n",
    "            tmp_genome_graph.extend(sub_genome_graph)\n",
    "        genome_graph = tmp_genome_graph\n",
    "        \n",
    "    p = [] \n",
    "    \n",
    "    start_index = 0\n",
    "    \n",
    "    for i in range(start_index + 1, len(genome_graph)):\n",
    "        if (i % 2 == 0) & (abs(genome_graph[i] - genome_graph[i - 1]) > 1):\n",
    "            nodes       = genome_graph[start_index : i]\n",
    "            nodes       = [nodes[-1]] + nodes[:-1]\n",
    "            choromosome = CycleToChromosome(nodes)\n",
    "            p.append(choromosome)\n",
    "            start_index = i\n",
    "            \n",
    "    nodes = genome_graph[start_index:]\n",
    "    nodes = [nodes[-1]] + nodes[:-1]\n",
    "    choromosome = CycleToChromosome(nodes)\n",
    "    p.append(choromosome)\n",
    "    \n",
    "    return(p)\n",
    "\n",
    "# Test\n",
    "genome_graph = '(2, 4), (3, 6), (5, 1), (7, 9), (10, 12), (11, 8)'\n",
    "\n",
    "GraphToGenome(genome_graph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Code Challenge: Solve the 2-Break Distance Problem.\n",
    "Input: Genomes P and Q.\n",
    "Output: The 2-break distance d(P, Q).\n",
    "'''\n",
    "\n",
    "def Two_Break_Distance(P, Q):\n",
    "    p = ColoredEdges(P)\n",
    "    q = ColoredEdges(Q)\n",
    "    \n",
    "    comb = p + q\n",
    "    \n",
    "    n_cycle = 0\n",
    "    while len(comb) != 0:\n",
    "        n_cycle = n_cycle + 1\n",
    "        current_cycle = comb[0]\n",
    "        comb.pop(0)\n",
    "        for number in current_cycle:\n",
    "            for edge in comb:\n",
    "                if number in edge:\n",
    "                    edge.remove(number)\n",
    "                    current_cycle.extend(edge)\n",
    "                    comb.remove(edge)\n",
    "    \n",
    "    return(len(q) - n_cycle)\n",
    "\n",
    "# Test\n",
    "P = '(+1 +2 +3 +4 +5 +6)'\n",
    "Q = '(+1 -3 -6 -5)(+2 -4)'\n",
    "\n",
    "Two_Break_Distance(P, Q)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[2, 4], [7, 5], [1, 3], [6, 8]]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Code Challenge: Implement 2-BreakOnGenomeGraph.\n",
    "Input: The colored edges of a genome graph GenomeGraph, followed by indices i1 , i2 , i3 , and i4 .\n",
    "Output: The colored edges of the genome graph resulting from applying the 2-break operation\n",
    "'''\n",
    "\n",
    "def Two_BreakOnGenomeGraph(genome_graph, i1 , i2 , i3 , i4):\n",
    "\n",
    "    if type(genome_graph) == str:\n",
    "        genome_graph = genome_graph[1: -1]\n",
    "        genome_graph = genome_graph.split('), (')\n",
    "        for i in range(len(genome_graph)):\n",
    "            genome_graph[i] = genome_graph[i].split(',')\n",
    "            genome_graph[i] = list(map(int, genome_graph[i]))\n",
    "\n",
    "    tmp_genome_graph = copy.deepcopy(genome_graph)\n",
    "    for edge in genome_graph:\n",
    "        \n",
    "        if (i1 in edge) & (i2 in edge):\n",
    "            tmp_genome_graph.remove(edge)\n",
    "            \n",
    "        \n",
    "        elif (i3 in edge) & (i4 in edge):\n",
    "            tmp_genome_graph.remove(edge)\n",
    "            \n",
    "    tmp_genome_graph.append([i1, i3])\n",
    "    tmp_genome_graph.append([i2, i4])\n",
    "    \n",
    "    return(tmp_genome_graph)\n",
    "\n",
    "# Test\n",
    "genome_graph = '(2, 4), (3, 8), (7, 5), (6, 1)'\n",
    "i1 , i2 , i3 , i4 = 1, 6, 3, 8\n",
    "\n",
    "Two_BreakOnGenomeGraph(genome_graph, i1 , i2 , i3 , i4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(+1 -2 -3 +4)\n",
      "(-3 +4 +1 +2)\n",
      "(-3 +4 -2 -1)\n",
      "(+4 -2 -1 +3)\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "2-Break Sorting Problem: Find a shortest transformation of one genome into another by 2-breaks.\n",
    "Input: Two genomes with circular chromosomes on the same set of synteny blocks.\n",
    "Output: The sequence of genomes resulting from applying a shortest sequence of 2-breaks\n",
    "    transforming one genome into the other.\n",
    "'''\n",
    "\n",
    "def Sort_Graphs(edges):\n",
    "    tmp_edges = copy.deepcopy(edges)\n",
    "    master_output_edges = []\n",
    "     \n",
    "    while len(tmp_edges) != 0:\n",
    "        output_edges = [copy.deepcopy(tmp_edges[0])]\n",
    "        tmp_edges.pop(0)\n",
    "        \n",
    "        while True:\n",
    "            if (output_edges[0][0]) % 2 == 0:\n",
    "                if (output_edges[0][0] - output_edges[-1][1]) == 1:\n",
    "                    break\n",
    "            if (output_edges[0][0]) % 2 == 1:\n",
    "                if (output_edges[0][0] - output_edges[-1][1]) == -1:\n",
    "                    break\n",
    "            \n",
    "            for edge in tmp_edges:\n",
    "\n",
    "                last_number  = output_edges[-1][1]\n",
    "                if (last_number % 2) == 0:\n",
    "                    if (last_number -1) in edge:\n",
    "                        if last_number - 1 == edge[0]:\n",
    "                            output_edges.append(edge)\n",
    "                        else:\n",
    "                            output_edges.append(edge[::-1])\n",
    "                        tmp_edges.remove(edge)\n",
    "                \n",
    "                else:\n",
    "                    if ( last_number + 1) in edge:\n",
    "                        if last_number + 1 == edge[0]:\n",
    "                            output_edges.append(edge)\n",
    "                        else:\n",
    "                            output_edges.append(edge[::-1])\n",
    "                        tmp_edges.remove(edge)\n",
    "\n",
    "        master_output_edges.extend(output_edges)\n",
    "    \n",
    "    return(master_output_edges)\n",
    "\n",
    "def print_genome(P):\n",
    "    output_print=''\n",
    "    tmp_P = copy.deepcopy(P)\n",
    "    \n",
    "    for sub_p in tmp_P:\n",
    "        print_sub = '('\n",
    "        \n",
    "        for i in range(len(sub_p)):\n",
    "            if sub_p[i] > 0:\n",
    "                sub_p[i] = '+' + str(sub_p[i])\n",
    "            else:\n",
    "                sub_p[i] = str(sub_p[i])\n",
    "                \n",
    "            print_sub = print_sub + sub_p[i] + ' '\n",
    "            \n",
    "        print_sub    = print_sub[: -1] + ')'\n",
    "        output_print = output_print + print_sub\n",
    "        \n",
    "    print(output_print)\n",
    "    \n",
    "\n",
    "def ShortestRearrangementScenario(P, Q):\n",
    "    print(P)\n",
    "    red_edges  = ColoredEdges(P)\n",
    "    blue_edges = ColoredEdges(Q)\n",
    "    \n",
    "    red_blue_edges = red_edges + blue_edges\n",
    "    \n",
    "    while Two_Break_Distance(P, Q) != 0:\n",
    "        \n",
    "        for blue_edge in blue_edges:\n",
    "            if (blue_edge[::-1] not in red_edges) & (blue_edge not in red_edges):\n",
    "                i1, i3 = blue_edge[0], blue_edge[1]\n",
    "                for red_edge in red_edges:\n",
    "                    if red_edge[0] == i1:\n",
    "                        i2 = red_edge[1]\n",
    "                    elif red_edge[1] == i1:\n",
    "                        i2 = red_edge[0]\n",
    "                    \n",
    "                    if red_edge[0] == i3:\n",
    "                        i4 = red_edge[1]\n",
    "                    elif red_edge[1] == i3:\n",
    "                        i4 = red_edge[0]\n",
    "                \n",
    "                red_edges = Two_BreakOnGenomeGraph(red_edges, i1 , i2 , i3 , i4)\n",
    "                red_edges = Sort_Graphs(red_edges)\n",
    "                P         = GraphToGenome(red_edges)\n",
    "                print_genome(P)\n",
    "                \n",
    "                break\n",
    "\n",
    "# Test \n",
    "P = '(+1 -2 -3 +4)'\n",
    "Q = '(+1 +2 -4 -3)'\n",
    "\n",
    "ShortestRearrangementScenario(P, Q)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[0, 4], [0, 0], [4, 2], [6, 6]]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "Shared k-mers Problem: Given two strings, find all their shared k-mers.\n",
    "Input: An integer k and two strings.\n",
    "Output: All k-mers shared by these strings, in the form of ordered pairs (x, y) corresponding to starting positions\n",
    "     of these k-mers in the respective strings.\n",
    "'''\n",
    "\n",
    "def ReverseComplement(DNA_string):\n",
    "    complementry_rule = {\n",
    "        'A' : 'T',\n",
    "        'T' : 'A',\n",
    "        'C' : 'G',\n",
    "        'G' : 'C' \n",
    "    }\n",
    "    com_string = ''\n",
    "    for base in DNA_string:\n",
    "        com_string = com_string + complementry_rule[base]\n",
    "    \n",
    "    com_string = com_string[::-1]\n",
    "    \n",
    "    return (com_string)\n",
    "\n",
    "def Shared_Kmers(string1, string2, k):\n",
    "    kmers_position_1 = dict()\n",
    "    kmers_position_2 = dict()\n",
    "    shared_positions = []\n",
    "    \n",
    "    for start in range(len(string1) - k + 1):\n",
    "        end     = start + k\n",
    "        pattern = string1[start : end]\n",
    "        if pattern in kmers_position_1:\n",
    "            kmers_position_1[pattern].append(start)\n",
    "        else:\n",
    "            kmers_position_1[pattern] = [start]\n",
    "    \n",
    "    for start in range(len(string2) - k + 1):\n",
    "        end     = start + k\n",
    "        pattern = string2[start : end]\n",
    "        if pattern in kmers_position_2:\n",
    "            kmers_position_2[pattern].append(start)\n",
    "        else:\n",
    "            kmers_position_2[pattern] = [start]\n",
    "    \n",
    "    for pattern, positions in kmers_position_1.items():\n",
    "        \n",
    "        if pattern in kmers_position_2:\n",
    "\n",
    "            for position_1 in positions:\n",
    "                for position_2 in kmers_position_2[pattern]:\n",
    "                    shared_positions.append([position_1, position_2])\n",
    "        \n",
    "        if ReverseComplement(pattern) in kmers_position_2:\n",
    "\n",
    "            for position_1 in positions:\n",
    "                for position_2 in kmers_position_2[ReverseComplement(pattern)]:\n",
    "                    shared_positions.append([position_1, position_2])\n",
    "        \n",
    "    return(shared_positions)\n",
    "\n",
    "# Test\n",
    "k = 3\n",
    "string1 = 'AAACTCATC'\n",
    "string2 = 'TTTCAAATC'\n",
    "\n",
    "Shared_Kmers(string1, string2, k)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
